// @twilio-paste/data-visualization-library/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// ../../../node_modules/deepmerge/dist/cjs.js
var require_cjs = __commonJS({
  "../../../node_modules/deepmerge/dist/cjs.js"(exports, module2) {
    "use strict";
    var isMergeableObject = function isMergeableObject2(value) {
      return isNonNullObject(value) && !isSpecial(value);
    };
    function isNonNullObject(value) {
      return !!value && typeof value === "object";
    }
    function isSpecial(value) {
      var stringValue = Object.prototype.toString.call(value);
      return stringValue === "[object RegExp]" || stringValue === "[object Date]" || isReactElement(value);
    }
    var canUseSymbol = typeof Symbol === "function" && Symbol.for;
    var REACT_ELEMENT_TYPE = canUseSymbol ? Symbol.for("react.element") : 60103;
    function isReactElement(value) {
      return value.$$typeof === REACT_ELEMENT_TYPE;
    }
    function emptyTarget(val) {
      return Array.isArray(val) ? [] : {};
    }
    function cloneUnlessOtherwiseSpecified(value, options) {
      return options.clone !== false && options.isMergeableObject(value) ? deepmerge(emptyTarget(value), value, options) : value;
    }
    function defaultArrayMerge(target, source, options) {
      return target.concat(source).map(function(element) {
        return cloneUnlessOtherwiseSpecified(element, options);
      });
    }
    function getMergeFunction(key, options) {
      if (!options.customMerge) {
        return deepmerge;
      }
      var customMerge = options.customMerge(key);
      return typeof customMerge === "function" ? customMerge : deepmerge;
    }
    function getEnumerableOwnPropertySymbols(target) {
      return Object.getOwnPropertySymbols ? Object.getOwnPropertySymbols(target).filter(function(symbol) {
        return target.propertyIsEnumerable(symbol);
      }) : [];
    }
    function getKeys(target) {
      return Object.keys(target).concat(getEnumerableOwnPropertySymbols(target));
    }
    function propertyIsOnObject(object, property) {
      try {
        return property in object;
      } catch (_) {
        return false;
      }
    }
    function propertyIsUnsafe(target, key) {
      return propertyIsOnObject(target, key) && !(Object.hasOwnProperty.call(target, key) && Object.propertyIsEnumerable.call(target, key));
    }
    function mergeObject(target, source, options) {
      var destination = {};
      if (options.isMergeableObject(target)) {
        getKeys(target).forEach(function(key) {
          destination[key] = cloneUnlessOtherwiseSpecified(target[key], options);
        });
      }
      getKeys(source).forEach(function(key) {
        if (propertyIsUnsafe(target, key)) {
          return;
        }
        if (propertyIsOnObject(target, key) && options.isMergeableObject(source[key])) {
          destination[key] = getMergeFunction(key, options)(target[key], source[key], options);
        } else {
          destination[key] = cloneUnlessOtherwiseSpecified(source[key], options);
        }
      });
      return destination;
    }
    function deepmerge(target, source, options) {
      options = options || {};
      options.arrayMerge = options.arrayMerge || defaultArrayMerge;
      options.isMergeableObject = options.isMergeableObject || isMergeableObject;
      options.cloneUnlessOtherwiseSpecified = cloneUnlessOtherwiseSpecified;
      var sourceIsArray = Array.isArray(source);
      var targetIsArray = Array.isArray(target);
      var sourceAndTargetTypesMatch = sourceIsArray === targetIsArray;
      if (!sourceAndTargetTypesMatch) {
        return cloneUnlessOtherwiseSpecified(source, options);
      } else if (sourceIsArray) {
        return options.arrayMerge(target, source, options);
      } else {
        return mergeObject(target, source, options);
      }
    }
    deepmerge.all = function deepmergeAll(array, options) {
      if (!Array.isArray(array)) {
        throw new Error("first argument should be an array");
      }
      return array.reduce(function(prev, next) {
        return deepmerge(prev, next, options);
      }, {});
    };
    var deepmerge_1 = deepmerge;
    module2.exports = deepmerge_1;
  }
});

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  usePasteHighchartsTheme: () => usePasteHighchartsTheme
});
module.exports = __toCommonJS(src_exports);

// src/usePasteHighchartsTheme.tsx
var React = __toESM(require("react"));
var import_styling_library = require("@twilio-paste/styling-library");
var import_deepmerge = __toESM(require_cjs());
var usePasteHighchartsTheme = (options) => {
  if (options == null) {
    throw new Error(
      `[usePasteHighchartsTheme]: Must provide highcharts options into this function. Options are deepmerged against the theme and returned for usage.`
    );
  }
  const context = React.useContext(import_styling_library.ThemeContext);
  if (!context) {
    throw new Error(
      "[usePasteHighchartsTheme]: must be used within the @twilio-paste/theme provider. https://paste.twilio.design/introduction/for-engineers/manual-installation#setting-up-the-theme-provider"
    );
  }
  const colors = React.useMemo(() => {
    const collator = new Intl.Collator(void 0, { numeric: true, sensitivity: "base" });
    const tokenNames = Object.keys(context.dataVisualization);
    const sortedTokenNames = tokenNames.sort((a, b) => collator.compare(a, b));
    return sortedTokenNames.map((name) => context.dataVisualization[name]);
  }, [context]);
  return (0, import_deepmerge.default)(options, {
    colors,
    chart: {
      backgroundColor: context.backgroundColors.colorBackgroundBody
    },
    title: {
      style: {
        color: context.textColors.colorText,
        font: `${context.fontWeights.fontWeightBold} ${context.fontSizes.fontSize40} ${context.fonts.fontFamilyText}`
      }
    },
    subtitle: {
      style: {
        color: context.textColors.colorTextWeak,
        font: `${context.fontSizes.fontSize20} ${context.fonts.fontFamilyText}`
      }
    },
    legend: {
      itemStyle: {
        font: `${context.fontSizes.fontSize10} ${context.fonts.fontFamilyText}`,
        color: context.textColors.colorText
      },
      itemHoverStyle: {
        color: context.textColors.colorTextWeak
      },
      title: {
        style: {
          color: context.textColors.colorText
        }
      }
    },
    xAxis: {
      lineColor: context.textColors.colorText,
      tickColor: context.textColors.colorText,
      labels: {
        style: {
          color: context.textColors.colorText,
          font: `${context.fontWeights.fontWeightMedium} ${context.fontSizes.fontSize10} ${context.fonts.fontFamilyText}`
        }
      },
      title: {
        style: {
          color: context.textColors.colorText,
          font: `${context.fontWeights.fontWeightSemibold} ${context.fontSizes.fontSize30} ${context.fonts.fontFamilyText}`
        }
      }
    },
    yAxis: {
      labels: {
        style: {
          color: context.textColors.colorText,
          font: `${context.fontWeights.fontWeightMedium} ${context.fontSizes.fontSize10} ${context.fonts.fontFamilyText}`
        }
      },
      title: {
        style: {
          color: context.textColors.colorText,
          font: `${context.fontWeights.fontWeightSemibold} ${context.fontSizes.fontSize30} ${context.fonts.fontFamilyText}`
        }
      }
    },
    labels: {
      style: {
        color: context.textColors.colorText
      }
    },
    tooltip: {
      backgroundColor: context.backgroundColors.colorBackground,
      style: {
        color: context.textColors.colorText
      }
    },
    credits: {
      style: {
        color: context.textColors.colorTextWeaker
      }
    },
    plotOptions: {
      series: {
        borderColor: "transparent"
      }
    }
  });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  usePasteHighchartsTheme
});
