// @twilio-paste/side-modal/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  SideModal: () => SideModal,
  SideModalBody: () => SideModalBody,
  SideModalButton: () => SideModalButton,
  SideModalContainer: () => SideModalContainer,
  SideModalFooter: () => SideModalFooter,
  SideModalFooterActions: () => SideModalFooterActions,
  SideModalHeader: () => SideModalHeader,
  SideModalHeading: () => SideModalHeading,
  useSideModalState: () => import_non_modal_dialog_primitive4.useNonModalDialogPrimitiveState
});
module.exports = __toCommonJS(src_exports);

// src/SideModal.tsx
var React2 = __toESM(require("react"));
var import_prop_types = __toESM(require("prop-types"));
var import_box = require("@twilio-paste/box");
var import_theme = require("@twilio-paste/theme");
var import_non_modal_dialog_primitive = require("@twilio-paste/non-modal-dialog-primitive");

// src/SideModalContext.tsx
var React = __toESM(require("react"));
var SideModalContext = React.createContext({});

// src/SideModal.tsx
var StyledSideModal = React2.forwardRef((_a, ref) => {
  var _b = _a, { style } = _b, props = __objRest(_b, ["style"]);
  return /* @__PURE__ */ React2.createElement(
    import_box.Box,
    __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
      ref,
      style,
      boxShadow: "shadow",
      width: "size60",
      zIndex: "zIndex80",
      transform: "none!important",
      inset: "0rem 0rem 0rem auto!important",
      _focus: {
        outline: "none"
      }
    })
  );
});
StyledSideModal.displayName = "StyledSideModal";
var SideModal = React2.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "SIDE_MODAL" } = _b, props = __objRest(_b, ["children", "element"]);
    const dialog = React2.useContext(SideModalContext);
    return /* @__PURE__ */ React2.createElement(
      import_non_modal_dialog_primitive.NonModalDialogPrimitive,
      __spreadProps(__spreadValues(__spreadValues({}, dialog), (0, import_box.safelySpreadBoxProps)(props)), {
        as: StyledSideModal,
        element: `${element}_CONTAINER`,
        ref,
        preventBodyScroll: false,
        hideOnClickOutside: false
      }),
      /* @__PURE__ */ React2.createElement(import_theme.StyledBase, null, /* @__PURE__ */ React2.createElement(
        import_box.Box,
        {
          element,
          display: "grid",
          gridTemplateRows: "auto 1fr auto",
          height: "100vh",
          backgroundColor: "colorBackgroundBody"
        },
        children
      ))
    );
  }
);
SideModal.displayName = "SideModal";
SideModal.propTypes = {
  "aria-label": import_prop_types.default.string.isRequired,
  children: import_prop_types.default.node.isRequired,
  element: import_prop_types.default.string
};

// src/SideModalButton.tsx
var React3 = __toESM(require("react"));
var import_prop_types2 = __toESM(require("prop-types"));
var import_button = require("@twilio-paste/button");
var import_non_modal_dialog_primitive2 = require("@twilio-paste/non-modal-dialog-primitive");
var SideModalButton = React3.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "SIDE_MODAL_BUTTON" } = _b, props = __objRest(_b, ["children", "element"]);
    const dialog = React3.useContext(SideModalContext);
    return /* @__PURE__ */ React3.createElement(import_non_modal_dialog_primitive2.NonModalDialogDisclosurePrimitive, __spreadProps(__spreadValues(__spreadValues({ element }, dialog), props), { as: import_button.Button, ref }), children);
  }
);
SideModalButton.propTypes = {
  id: import_prop_types2.default.string,
  children: import_prop_types2.default.node.isRequired,
  element: import_prop_types2.default.string
};
SideModalButton.displayName = "SideModalButton";

// src/SideModalHeader.tsx
var React4 = __toESM(require("react"));
var import_prop_types3 = __toESM(require("prop-types"));
var import_box2 = require("@twilio-paste/box");
var import_button2 = require("@twilio-paste/button");
var import_CloseIcon = require("@twilio-paste/icons/cjs/CloseIcon");
var import_screen_reader_only = require("@twilio-paste/screen-reader-only");
var SideModalHeader = React4.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "SIDE_MODAL_HEADER", i18nDismissLabel = "close" } = _b, props = __objRest(_b, ["children", "element", "i18nDismissLabel"]);
    const { hide } = React4.useContext(SideModalContext);
    return /* @__PURE__ */ React4.createElement(
      import_box2.Box,
      __spreadProps(__spreadValues({}, (0, import_box2.safelySpreadBoxProps)(props)), {
        ref,
        paddingRight: "space90",
        paddingLeft: "space90",
        paddingTop: "space90",
        element,
        display: "flex",
        flexDirection: "row"
      }),
      /* @__PURE__ */ React4.createElement(import_box2.Box, { display: "flex", flexGrow: 1, marginRight: "space70", alignItems: "center" }, children),
      /* @__PURE__ */ React4.createElement(import_button2.Button, { element: `${element}_CLOSE_BUTTON`, variant: "secondary_icon", size: "reset", onClick: hide }, /* @__PURE__ */ React4.createElement(import_CloseIcon.CloseIcon, { element: `${element}_CLOSE_ICON`, decorative: true, size: "sizeIcon20" }), /* @__PURE__ */ React4.createElement(import_screen_reader_only.ScreenReaderOnly, null, i18nDismissLabel))
    );
  }
);
SideModalHeader.displayName = "SideModalHeader";
SideModalHeader.propTypes = {
  children: import_prop_types3.default.node,
  i18nDismissLabel: import_prop_types3.default.string,
  element: import_prop_types3.default.string
};

// src/SideModalHeading.tsx
var React5 = __toESM(require("react"));
var import_prop_types4 = __toESM(require("prop-types"));
var import_heading = require("@twilio-paste/heading");
var SideModalHeading = React5.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, as = "h2", element = "SIDE_MODAL_HEADING" } = _b, props = __objRest(_b, ["children", "as", "element"]);
    return /* @__PURE__ */ React5.createElement(import_heading.Heading, __spreadProps(__spreadValues({}, props), { as, element, marginBottom: "space0", variant: "heading30", ref }), children);
  }
);
SideModalHeading.displayName = "SideModalHeading";
SideModalHeading.propTypes = {
  children: import_prop_types4.default.node.isRequired,
  as: import_prop_types4.default.oneOf(["h1", "h2", "h3", "h4", "h5", "h6"]),
  element: import_prop_types4.default.string
};

// src/SideModalContainer.tsx
var React6 = __toESM(require("react"));
var import_non_modal_dialog_primitive3 = require("@twilio-paste/non-modal-dialog-primitive");
var import_non_modal_dialog_primitive4 = require("@twilio-paste/non-modal-dialog-primitive");
var BaseSideModalContainer = (_a) => {
  var _b = _a, {
    gutter,
    children,
    placement,
    modal,
    state
  } = _b, initialState = __objRest(_b, [
    "gutter",
    "children",
    "placement",
    "modal",
    "state"
  ]);
  const dialog = state || (0, import_non_modal_dialog_primitive3.useNonModalDialogPrimitiveState)(__spreadValues({
    modal: true
  }, initialState));
  return /* @__PURE__ */ React6.createElement(SideModalContext.Provider, { value: __spreadValues({}, dialog) }, children);
};
BaseSideModalContainer.displayName = "BaseSideModalContainer";
var SideModalContainer = React6.memo(BaseSideModalContainer);
SideModalContainer.displayName = "SideModalContainer";

// src/SideModalBody.tsx
var React7 = __toESM(require("react"));
var import_prop_types5 = __toESM(require("prop-types"));
var import_box3 = require("@twilio-paste/box");
var SideModalBody = React7.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "SIDE_MODAL_BODY" } = _b, props = __objRest(_b, ["children", "element"]);
    return /* @__PURE__ */ React7.createElement(import_box3.Box, __spreadProps(__spreadValues({}, (0, import_box3.safelySpreadBoxProps)(props)), { element, ref, overflowY: "auto", padding: "space90" }), children);
  }
);
SideModalBody.propTypes = {
  children: import_prop_types5.default.node,
  element: import_prop_types5.default.string
};
SideModalBody.displayName = "SideModalBody";

// src/SideModalFooter.tsx
var React8 = __toESM(require("react"));
var import_prop_types6 = __toESM(require("prop-types"));
var import_box4 = require("@twilio-paste/box");
var SideModalFooter = React8.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "SIDE_MODAL_FOOTER" } = _b, props = __objRest(_b, ["children", "element"]);
    return /* @__PURE__ */ React8.createElement(
      import_box4.Box,
      __spreadProps(__spreadValues({}, (0, import_box4.safelySpreadBoxProps)(props)), {
        as: "div",
        element,
        ref,
        paddingRight: "space90",
        paddingLeft: "space90",
        paddingBottom: "space90",
        display: "flex"
      }),
      children
    );
  }
);
SideModalFooter.displayName = "SideModalFooter";
SideModalFooter.propTypes = {
  children: import_prop_types6.default.node.isRequired,
  element: import_prop_types6.default.string
};

// src/SideModalFooterActions.tsx
var React9 = __toESM(require("react"));
var import_prop_types7 = __toESM(require("prop-types"));
var import_box5 = require("@twilio-paste/box");
var SideModalFooterActions = React9.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "SIDE_MODAL_FOOTER_ACTIONS", justifyContent = "flex-end" } = _b, props = __objRest(_b, ["children", "element", "justifyContent"]);
    return /* @__PURE__ */ React9.createElement(
      import_box5.Box,
      __spreadProps(__spreadValues({}, (0, import_box5.safelySpreadBoxProps)(props)), {
        display: "flex",
        justifyContent,
        flexWrap: "wrap",
        flexGrow: 1,
        element,
        columnGap: "space50",
        alignItems: "center",
        ref
      }),
      children
    );
  }
);
SideModalFooterActions.displayName = "SideModalFooterActions";
SideModalFooterActions.propTypes = {
  children: import_prop_types7.default.node.isRequired,
  element: import_prop_types7.default.string,
  justifyContent: import_prop_types7.default.oneOf(["flex-start", "flex-end", "space-between"])
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SideModal,
  SideModalBody,
  SideModalButton,
  SideModalContainer,
  SideModalFooter,
  SideModalFooterActions,
  SideModalHeader,
  SideModalHeading,
  useSideModalState
});
