// @twilio-paste/list/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  List: () => List,
  ListItem: () => ListItem,
  OrderedList: () => OrderedList,
  UnorderedList: () => UnorderedList
});
module.exports = __toCommonJS(src_exports);

// src/List.tsx
var React = __toESM(require("react"));
var import_prop_types = __toESM(require("prop-types"));
var import_text = require("@twilio-paste/text");
var import_style_props = require("@twilio-paste/style-props");
var List = React.forwardRef(
  (_a, ref) => {
    var _b = _a, { as, children, element = "LIST", listStyleType = "disc" } = _b, props = __objRest(_b, ["as", "children", "element", "listStyleType"]);
    return /* @__PURE__ */ React.createElement(
      import_text.Text,
      __spreadProps(__spreadValues({}, props), {
        as,
        color: "colorText",
        element,
        fontSize: "fontSize30",
        fontWeight: "fontWeightNormal",
        lineHeight: "lineHeight40",
        marginLeft: "space70",
        listStyleType,
        ref
      }),
      children
    );
  }
);
List.displayName = "List";
List.propTypes = {
  as: import_prop_types.default.oneOf(["ol", "ul"]).isRequired,
  element: import_prop_types.default.string,
  marginTop: import_style_props.isMarginTokenProp,
  marginBottom: import_style_props.isMarginTokenProp
};

// src/ListItem.tsx
var React2 = __toESM(require("react"));
var import_prop_types2 = __toESM(require("prop-types"));
var import_text2 = require("@twilio-paste/text");
var ListItem = React2.forwardRef((_a, ref) => {
  var _b = _a, { children, element = "LIST_ITEM" } = _b, props = __objRest(_b, ["children", "element"]);
  return /* @__PURE__ */ React2.createElement(
    import_text2.Text,
    __spreadProps(__spreadValues({}, (0, import_text2.safelySpreadTextProps)(props)), {
      as: "li",
      color: "colorText",
      element,
      fontSize: "fontSize30",
      fontWeight: "fontWeightNormal",
      lineHeight: "lineHeight40",
      marginBottom: "space30",
      marginTop: "space30",
      ref
    }),
    children
  );
});
ListItem.displayName = "ListItem";
ListItem.propTypes = {
  element: import_prop_types2.default.string
};

// src/OrderedList.tsx
var React3 = __toESM(require("react"));
var import_prop_types3 = __toESM(require("prop-types"));
var import_text3 = require("@twilio-paste/text");
var import_style_props2 = require("@twilio-paste/style-props");
var OrderedList = React3.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "ORDERED_LIST", marginTop, marginBottom = "space70" } = _b, props = __objRest(_b, ["children", "element", "marginTop", "marginBottom"]);
    return /* @__PURE__ */ React3.createElement(
      List,
      __spreadProps(__spreadValues({}, (0, import_text3.safelySpreadTextProps)(props)), {
        as: "ol",
        element,
        marginTop,
        marginBottom,
        listStyleType: "decimal",
        ref
      }),
      children
    );
  }
);
OrderedList.displayName = "OrderedList";
OrderedList.propTypes = {
  element: import_prop_types3.default.string,
  marginTop: import_style_props2.isMarginTokenProp,
  marginBottom: import_style_props2.isMarginTokenProp
};

// src/UnorderedList.tsx
var React4 = __toESM(require("react"));
var import_prop_types4 = __toESM(require("prop-types"));
var import_text4 = require("@twilio-paste/text");
var import_style_props3 = require("@twilio-paste/style-props");
var UnorderedList = React4.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "UNORDERED_LIST", marginTop, marginBottom = "space70" } = _b, props = __objRest(_b, ["children", "element", "marginTop", "marginBottom"]);
    return /* @__PURE__ */ React4.createElement(
      List,
      __spreadProps(__spreadValues({}, (0, import_text4.safelySpreadTextProps)(props)), {
        as: "ul",
        element,
        marginTop,
        marginBottom,
        ref,
        listStyleType: "disc"
      }),
      children
    );
  }
);
UnorderedList.displayName = "UnorderedList";
UnorderedList.propTypes = {
  element: import_prop_types4.default.string,
  marginTop: import_style_props3.isMarginTokenProp,
  marginBottom: import_style_props3.isMarginTokenProp
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  List,
  ListItem,
  OrderedList,
  UnorderedList
});
