// @twilio-paste/menu-primitive/dist/index.debug.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  MenuPrimitive: () => import_reakit_library.Menu,
  MenuPrimitiveButton: () => import_reakit_library.MenuButton,
  MenuPrimitiveGroup: () => import_reakit_library.MenuGroup,
  MenuPrimitiveItem: () => import_reakit_library.MenuItem,
  MenuPrimitiveItemCheckbox: () => import_reakit_library.MenuItemCheckbox,
  MenuPrimitiveItemRadioMenuItemRadio: () => import_reakit_library.MenuItemRadio,
  MenuPrimitiveSeparator: () => import_reakit_library.MenuSeparator,
  useMenuPrimitiveState: () => import_reakit_library.useMenuState
});
module.exports = __toCommonJS(src_exports);
var import_reakit_library = require("@twilio-paste/reakit-library");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MenuPrimitive,
  MenuPrimitiveButton,
  MenuPrimitiveGroup,
  MenuPrimitiveItem,
  MenuPrimitiveItemCheckbox,
  MenuPrimitiveItemRadioMenuItemRadio,
  MenuPrimitiveSeparator,
  useMenuPrimitiveState
});
