// @twilio-paste/table/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  TBody: () => TBody,
  TBodyPropTypes: () => TBodyPropTypes,
  TFoot: () => TFoot,
  TFootPropTypes: () => TFootPropTypes,
  THead: () => THead,
  THeadPropTypes: () => THeadPropTypes,
  Table: () => Table,
  TablePropTypes: () => TablePropTypes,
  Td: () => Td,
  TdPropTypes: () => TdPropTypes,
  Th: () => Th,
  ThPropTypes: () => ThPropTypes,
  Tr: () => Tr,
  TrPropTypes: () => TrPropTypes
});
module.exports = __toCommonJS(src_exports);

// src/Table.tsx
var React2 = __toESM(require("react"));
var import_box = require("@twilio-paste/box");

// src/TableContext.tsx
var React = __toESM(require("react"));
var TableContext = React.createContext({ striped: false });

// src/proptypes.ts
var import_prop_types = __toESM(require("prop-types"));
var import_style_props = require("@twilio-paste/style-props");

// src/constants.ts
var TableAlignmentObject = {
  LEFT: "left",
  CENTER: "center",
  RIGHT: "right"
};
var TableLayoutObject = {
  AUTO: "auto",
  FIXED: "fixed"
};
var TableVariantObject = {
  DEFAULT: "default",
  BORDERLESS: "borderless"
};
var TableVerticalAlignmentObject = {
  TOP: "top",
  MIDDLE: "middle",
  BOTTOM: "bottom"
};

// src/proptypes.ts
var TablePropTypes = {
  children: import_prop_types.default.node.isRequired,
  element: import_prop_types.default.string,
  isActionable: import_prop_types.default.bool,
  scrollHorizontally: import_prop_types.default.bool,
  noWrap: import_prop_types.default.bool,
  striped: import_prop_types.default.bool,
  tableLayout: import_prop_types.default.oneOf(Object.values(TableLayoutObject)),
  variant: import_prop_types.default.oneOf(Object.values(TableVariantObject))
};
var THeadPropTypes = {
  children: import_prop_types.default.node.isRequired,
  element: import_prop_types.default.string
};
var TBodyPropTypes = {
  children: import_prop_types.default.node.isRequired,
  element: import_prop_types.default.string
};
var TrPropTypes = {
  children: import_prop_types.default.node.isRequired,
  element: import_prop_types.default.string,
  verticalAlign: import_prop_types.default.oneOf(Object.values(TableVerticalAlignmentObject))
};
var ThPropTypes = {
  children: import_prop_types.default.node,
  element: import_prop_types.default.string,
  textAlign: import_prop_types.default.oneOf(Object.values(TableAlignmentObject)),
  width: import_style_props.isWidthTokenProp
};
var TdPropTypes = {
  children: import_prop_types.default.node,
  element: import_prop_types.default.string,
  textAlign: import_prop_types.default.oneOf(Object.values(TableAlignmentObject))
};
var TFootPropTypes = {
  children: import_prop_types.default.node.isRequired,
  element: import_prop_types.default.string
};

// src/Table.tsx
var Table = React2.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      element = "TABLE",
      id,
      isActionable,
      scrollHorizontally,
      noWrap,
      striped = false,
      tableLayout = "auto",
      variant = "default"
    } = _b, props = __objRest(_b, [
      "element",
      "id",
      "isActionable",
      "scrollHorizontally",
      "noWrap",
      "striped",
      "tableLayout",
      "variant"
    ]);
    const tableContext = {
      striped
    };
    return /* @__PURE__ */ React2.createElement(TableContext.Provider, { value: tableContext }, /* @__PURE__ */ React2.createElement(
      import_box.Box,
      {
        id,
        element: `${element}_WRAPPER`,
        overflowX: scrollHorizontally ? "auto" : null,
        whiteSpace: noWrap ? "nowrap" : null,
        boxShadow: isActionable ? "shadowFocus" : null
      },
      /* @__PURE__ */ React2.createElement(
        import_box.Box,
        __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
          ref,
          as: "table",
          borderCollapse: "separate",
          borderColor: "colorBorderWeaker",
          borderSpacing: "0",
          borderStyle: "solid",
          borderWidth: variant === "borderless" ? "borderWidth0" : "borderWidth10",
          element,
          tableLayout: tableLayout === "fixed" ? "fixed" : "auto",
          width: "100%"
        })
      )
    ));
  }
);
Table.displayName = "Table";
Table.propTypes = TablePropTypes;

// src/THead.tsx
var React3 = __toESM(require("react"));
var import_box2 = require("@twilio-paste/box");
var import_styling_library = require("@twilio-paste/styling-library");
var getStickyStyles = ({ stickyHeader }) => {
  if (!stickyHeader)
    return {};
  return {
    position: "sticky",
    zIndex: "zIndex10"
  };
};
var StyledTHead = import_styling_library.styled.thead(
  (0, import_styling_library.css)({
    backgroundColor: "colorBackground",
    top: (props) => props.top,
    "& > tr > th": {
      borderBottomWidth: "borderWidth20"
    }
  })
);
var THead = React3.forwardRef(
  (_a, ref) => {
    var _b = _a, { element = "THEAD", stickyHeader = false, top = stickyHeader ? "-1px" : null } = _b, props = __objRest(_b, ["element", "stickyHeader", "top"]);
    return /* @__PURE__ */ React3.createElement(
      import_box2.Box,
      __spreadValues(__spreadProps(__spreadValues({}, (0, import_box2.safelySpreadBoxProps)(props)), {
        ref,
        as: StyledTHead,
        element,
        top
      }), getStickyStyles({ stickyHeader }))
    );
  }
);
THead.displayName = "THead";
THead.propTypes = THeadPropTypes;

// src/TBody.tsx
var React4 = __toESM(require("react"));
var import_box3 = require("@twilio-paste/box");
var import_styling_library2 = require("@twilio-paste/styling-library");
var StyledTBody = import_styling_library2.styled.tbody(
  (0, import_styling_library2.css)({
    backgroundColor: "colorBackgroundBody",
    "& > tr:last-of-type > td": {
      borderBottomWidth: "borderWidth0"
    }
  })
);
var TBody = React4.forwardRef((_a, ref) => {
  var _b = _a, { element = "TBODY", children } = _b, props = __objRest(_b, ["element", "children"]);
  return /* @__PURE__ */ React4.createElement(import_box3.Box, __spreadProps(__spreadValues({ as: StyledTBody }, (0, import_box3.safelySpreadBoxProps)(props)), { ref, element }), children);
});
TBody.displayName = "TBody";
TBody.propTypes = TBodyPropTypes;

// src/TFoot.tsx
var React5 = __toESM(require("react"));
var import_box4 = require("@twilio-paste/box");
var import_styling_library3 = require("@twilio-paste/styling-library");
var StyledTFoot = import_styling_library3.styled.tfoot(
  (0, import_styling_library3.css)({
    "& > tr > td": {
      borderBottomWidth: "borderWidth0",
      borderTopWidth: "borderWidth20"
    },
    "& > tr > th": {
      borderBottomWidth: "borderWidth0"
    }
  })
);
var TFoot = React5.forwardRef((_a, ref) => {
  var _b = _a, { element = "TFOOT" } = _b, props = __objRest(_b, ["element"]);
  return /* @__PURE__ */ React5.createElement(
    import_box4.Box,
    __spreadProps(__spreadValues({}, (0, import_box4.safelySpreadBoxProps)(props)), {
      ref,
      as: StyledTFoot,
      backgroundColor: "colorBackground",
      borderTopStyle: "solid",
      borderTopWidth: "borderWidth20",
      borderTopColor: "colorBorderWeaker",
      element
    })
  );
});
TFoot.displayName = "TFoot";
TFoot.propTypes = TFootPropTypes;

// src/Tr.tsx
var React6 = __toESM(require("react"));
var import_box5 = require("@twilio-paste/box");
var Tr = React6.forwardRef(
  (_a, ref) => {
    var _b = _a, { element = "TR", verticalAlign = "middle" } = _b, props = __objRest(_b, ["element", "verticalAlign"]);
    const { striped } = React6.useContext(TableContext);
    return /* @__PURE__ */ React6.createElement(
      import_box5.Box,
      __spreadProps(__spreadValues({}, (0, import_box5.safelySpreadBoxProps)(props)), {
        ref,
        as: "tr",
        element,
        verticalAlign,
        _even: { backgroundColor: striped ? "colorBackgroundRowStriped" : "transparent" },
        _last: { borderWidth: "borderWidth0" }
      })
    );
  }
);
Tr.displayName = "Tr";
Tr.propTypes = TrPropTypes;

// src/Td.tsx
var React7 = __toESM(require("react"));
var import_box6 = require("@twilio-paste/box");
var Td = React7.forwardRef((_a, ref) => {
  var _b = _a, { element = "TD", textAlign = "left" } = _b, props = __objRest(_b, ["element", "textAlign"]);
  return /* @__PURE__ */ React7.createElement(
    import_box6.Box,
    __spreadProps(__spreadValues({}, (0, import_box6.safelySpreadBoxProps)(props)), {
      ref,
      as: "td",
      borderStyle: "solid",
      borderColor: "colorBorderWeaker",
      borderWidth: "borderWidth0",
      borderBottomWidth: "borderWidth10",
      element,
      fontSize: "fontSize30",
      lineHeight: "lineHeight30",
      paddingBottom: "space40",
      paddingLeft: "space50",
      paddingRight: "space50",
      paddingTop: "space40",
      position: "relative",
      textAlign,
      verticalAlign: "inherit",
      wordWrap: "break-word"
    })
  );
});
Td.displayName = "Td";
Td.propTypes = TdPropTypes;

// src/Th.tsx
var React8 = __toESM(require("react"));
var import_box7 = require("@twilio-paste/box");
var Th = React8.forwardRef(
  (_a, ref) => {
    var _b = _a, { element = "TH", textAlign = "left", width } = _b, props = __objRest(_b, ["element", "textAlign", "width"]);
    return /* @__PURE__ */ React8.createElement(
      import_box7.Box,
      __spreadProps(__spreadValues({}, (0, import_box7.safelySpreadBoxProps)(props)), {
        ref,
        as: "th",
        borderBottomStyle: "solid",
        borderBottomColor: "colorBorderWeaker",
        borderBottomWidth: "borderWidth10",
        element,
        fontSize: "fontSize30",
        lineHeight: "lineHeight30",
        fontWeight: "fontWeightSemibold",
        paddingBottom: "space40",
        paddingLeft: "space50",
        paddingRight: "space50",
        paddingTop: "space40",
        position: "relative",
        textAlign,
        verticalAlign: "inherit",
        width,
        wordWrap: "break-word"
      })
    );
  }
);
Th.displayName = "Th";
Th.propTypes = ThPropTypes;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TBody,
  TBodyPropTypes,
  TFoot,
  TFootPropTypes,
  THead,
  THeadPropTypes,
  Table,
  TablePropTypes,
  Td,
  TdPropTypes,
  Th,
  ThPropTypes,
  Tr,
  TrPropTypes
});
