// @twilio-paste/media-object/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  MediaBody: () => MediaBody,
  MediaFigure: () => MediaFigure,
  MediaObject: () => MediaObject
});
module.exports = __toCommonJS(src_exports);

// src/MediaObject.tsx
var React = __toESM(require("react"));
var import_prop_types = __toESM(require("prop-types"));
var import_box = require("@twilio-paste/box");
var MediaObject = React.forwardRef(
  (_a, ref) => {
    var _b = _a, { as = "span", children, element = "MEDIA_OBJECT", marginBottom, marginTop, verticalAlign = "top" } = _b, props = __objRest(_b, ["as", "children", "element", "marginBottom", "marginTop", "verticalAlign"]);
    return /* @__PURE__ */ React.createElement(
      import_box.Box,
      __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
        alignItems: verticalAlign === "top" ? "flex-start" : "center",
        as,
        display: "flex",
        element,
        marginBottom,
        marginTop,
        ref
      }),
      children
    );
  }
);
MediaObject.displayName = "MediaObject";
MediaObject.propTypes = {
  children: import_prop_types.default.node.isRequired,
  verticalAlign: import_prop_types.default.oneOf(["center", "top"])
};

// src/MediaFigure.tsx
var React2 = __toESM(require("react"));
var import_prop_types2 = __toESM(require("prop-types"));
var import_box2 = require("@twilio-paste/box");
var import_style_props = require("@twilio-paste/style-props");
var MediaFigure = React2.forwardRef(
  (_a, ref) => {
    var _b = _a, { align = "start", as = "span", children, element = "MEDIA_FIGURE", spacing = "space0" } = _b, props = __objRest(_b, ["align", "as", "children", "element", "spacing"]);
    return /* @__PURE__ */ React2.createElement(
      import_box2.Box,
      __spreadProps(__spreadValues({}, (0, import_box2.safelySpreadBoxProps)(props)), {
        as,
        display: "flex",
        element,
        flexShrink: 0,
        marginLeft: align === "end" ? spacing : void 0,
        marginRight: align === "start" ? spacing : void 0,
        ref
      }),
      children
    );
  }
);
MediaFigure.displayName = "MediaFigure";
MediaFigure.propTypes = {
  align: import_prop_types2.default.oneOf(["start", "end"]),
  children: import_prop_types2.default.node.isRequired,
  spacing: import_style_props.isMarginTokenProp
};

// src/MediaBody.tsx
var React3 = __toESM(require("react"));
var import_prop_types3 = __toESM(require("prop-types"));
var import_box3 = require("@twilio-paste/box");
var MediaBody = React3.forwardRef(
  (_a, ref) => {
    var _b = _a, { as = "span", children, element = "MEDIA_BODY" } = _b, props = __objRest(_b, ["as", "children", "element"]);
    return /* @__PURE__ */ React3.createElement(import_box3.Box, __spreadProps(__spreadValues({}, (0, import_box3.safelySpreadBoxProps)(props)), { as, element, flex: 1, minWidth: "size0", ref }), children);
  }
);
MediaBody.displayName = "MediaBody";
MediaBody.propTypes = {
  children: import_prop_types3.default.node.isRequired
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MediaBody,
  MediaFigure,
  MediaObject
});
