// @twilio-paste/aspect-ratio/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  AspectRatio: () => AspectRatio
});
module.exports = __toCommonJS(src_exports);
var React = __toESM(require("react"));
var import_prop_types = require("prop-types");
var import_box = require("@twilio-paste/box");
var import_styling_library = require("@twilio-paste/styling-library");
var RATIO_REGEX = /^(\d+:\d*)$/;
var isCorrectPattern = (ratio) => RATIO_REGEX.test(ratio);
var handlePropValidation = ({ ratio }) => {
  const hasRatio = ratio != null && ratio !== "";
  if (!hasRatio) {
    throw new Error(`[Paste: AspectRatio] Missing 'ratio' prop.`);
  }
  if (!isCorrectPattern(ratio)) {
    throw new Error(`[Paste: AspectRatio] 'ratio' is invalid. Use a colon separated number pattern (4:3).`);
  }
};
var AspectRatioContainer = import_styling_library.styled.div`
  position: relative;

  embed,
  iframe,
  object,
  video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
  }
`;
var AspectRatio = React.forwardRef((props, ref) => {
  handlePropValidation(props);
  const aspectArray = props.ratio.split(":").map(Number);
  const aspectPercent = aspectArray[1] / aspectArray[0] * 100;
  return /* @__PURE__ */ React.createElement(AspectRatioContainer, { ref, style: { paddingBottom: `${aspectPercent}%` } }, /* @__PURE__ */ React.createElement(import_box.Box, { position: "absolute", top: 0, right: 0, bottom: 0, left: 0 }, props.children));
});
AspectRatio.displayName = "AspectRatio";
if (false) {
  AspectRatio.propTypes = {
    children: PropTypes.node.isRequired,
    ratio: PropTypes.string.isRequired
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AspectRatio
});
