// @twilio-paste/anchor/dist/index.debug.es.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};

// src/index.tsx
import * as React3 from "react";
import { Box as Box3, safelySpreadBoxProps } from "@twilio-paste/box";
import { LinkExternalIcon } from "@twilio-paste/icons/esm/LinkExternalIcon";

// src/DefaultAnchor.tsx
import * as React from "react";
import { Box } from "@twilio-paste/box";

// src/proptypes.ts
import PropTypes from "prop-types";
var AnchorPropTypes = {
  children: PropTypes.node.isRequired,
  href: PropTypes.string.isRequired,
  rel: PropTypes.string,
  showExternal: PropTypes.bool,
  tabIndex: PropTypes.oneOf([0, -1]),
  target: PropTypes.oneOf(["_self", "_blank", "_parent", "_top"]),
  variant: PropTypes.oneOf(["default", "inverse"]),
  element: PropTypes.string,
  i18nExternalLinkLabel: PropTypes.string
};

// src/DefaultAnchor.tsx
var DefaultAnchor = React.forwardRef((props, ref) => /* @__PURE__ */ React.createElement(
  Box,
  __spreadProps(__spreadValues({}, props), {
    as: "a",
    color: "colorTextLink",
    fontSize: "inherit",
    fontWeight: "inherit",
    lineHeight: "inherit",
    outline: "none",
    ref,
    textDecoration: "underline",
    _active: {
      color: "colorTextLinkStronger",
      textDecoration: "none"
    },
    _focus: {
      boxShadow: "shadowFocus",
      color: "colorTextLinkStronger",
      textDecoration: "none"
    },
    _hover: {
      color: "colorTextLinkStronger",
      textDecoration: "none"
    }
  }),
  props.children
));
DefaultAnchor.displayName = "DefaultAnchor";
if (false) {
  DefaultAnchor.propTypes = AnchorPropTypes;
}

// src/InverseAnchor.tsx
import * as React2 from "react";
import { Box as Box2 } from "@twilio-paste/box";
var InverseAnchor = React2.forwardRef((props, ref) => /* @__PURE__ */ React2.createElement(
  Box2,
  __spreadProps(__spreadValues({}, props), {
    as: "a",
    color: "colorTextInverse",
    fontSize: "inherit",
    fontWeight: "inherit",
    lineHeight: "inherit",
    outline: "none",
    ref,
    textDecoration: "underline",
    _active: {
      color: "colorTextInverse",
      textDecoration: "none"
    },
    _focus: {
      boxShadow: "shadowFocusInverse",
      color: "colorTextInverse",
      textDecoration: "none"
    },
    _hover: {
      color: "colorTextInverse",
      textDecoration: "none"
    }
  }),
  props.children
));
InverseAnchor.displayName = "InverseAnchor";
if (false) {
  InverseAnchor.propTypes = AnchorPropTypes;
}

// src/index.tsx
var AnchorVariants = {
  inverse: InverseAnchor,
  default: DefaultAnchor
};
var EXTERNAL_URL_REGEX = /^(https?:)\S*$/;
var EXTERNAL_TARGET_DEFAULT = "_blank";
var EXTERNAL_REL_DEFAULT = "noreferrer noopener";
var isExternalUrl = (url) => EXTERNAL_URL_REGEX.test(url);
var secureExternalLink = (href) => {
  if (!isExternalUrl(href))
    return void 0;
  return {
    rel: EXTERNAL_REL_DEFAULT,
    target: EXTERNAL_TARGET_DEFAULT
  };
};
var Anchor = React3.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      element = "ANCHOR",
      variant = "default",
      showExternal,
      display,
      height,
      minHeight,
      maxHeight,
      width,
      minWidth,
      maxWidth,
      size,
      margin,
      marginBottom,
      marginLeft,
      marginRight,
      marginTop,
      marginX = null,
      marginY = null,
      padding,
      paddingBottom,
      paddingLeft,
      paddingRight,
      paddingTop,
      paddingX = null,
      paddingY = null,
      verticalAlign,
      i18nExternalLinkLabel = "(link takes you to an external page)"
    } = _b, props = __objRest(_b, [
      "element",
      "variant",
      "showExternal",
      "display",
      "height",
      "minHeight",
      "maxHeight",
      "width",
      "minWidth",
      "maxWidth",
      "size",
      "margin",
      "marginBottom",
      "marginLeft",
      "marginRight",
      "marginTop",
      "marginX",
      "marginY",
      "padding",
      "paddingBottom",
      "paddingLeft",
      "paddingRight",
      "paddingTop",
      "paddingX",
      "paddingY",
      "verticalAlign",
      "i18nExternalLinkLabel"
    ]);
    const AnchorComponent = AnchorVariants[variant];
    return /* @__PURE__ */ React3.createElement(
      AnchorComponent,
      __spreadProps(__spreadValues(__spreadValues({
        href: props.href,
        ref,
        variant
      }, secureExternalLink(props.href)), safelySpreadBoxProps(props)), {
        element,
        display,
        height,
        minHeight,
        maxHeight,
        width,
        minWidth,
        maxWidth,
        size,
        margin,
        marginBottom,
        marginLeft,
        marginRight,
        marginTop,
        marginX,
        marginY,
        padding,
        paddingBottom,
        paddingLeft,
        paddingRight,
        paddingTop,
        paddingX,
        paddingY,
        verticalAlign
      }),
      showExternal ? /* @__PURE__ */ React3.createElement(Box3, { as: "span" }, props.children, /* @__PURE__ */ React3.createElement(Box3, { as: "span", display: "inline-block", flexShrink: 0, verticalAlign: "middle" }, /* @__PURE__ */ React3.createElement(LinkExternalIcon, { decorative: false, title: i18nExternalLinkLabel }))) : props.children
    );
  }
);
Anchor.displayName = "Anchor";
Anchor.propTypes = AnchorPropTypes;
export {
  Anchor,
  isExternalUrl,
  secureExternalLink
};
