// @twilio-paste/separator/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  Separator: () => Separator
});
module.exports = __toCommonJS(src_exports);
var React = __toESM(require("react"));
var import_prop_types = __toESM(require("prop-types"));
var import_box = require("@twilio-paste/box");
var import_style_props = require("@twilio-paste/style-props");
var Separator = React.forwardRef(
  (_a, ref) => {
    var _b = _a, { element = "SEPARATOR", orientation, horizontalSpacing, verticalSpacing } = _b, props = __objRest(_b, ["element", "orientation", "horizontalSpacing", "verticalSpacing"]);
    return /* @__PURE__ */ React.createElement(
      import_box.Box,
      __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
        "aria-orientation": orientation,
        element,
        margin: "space0",
        marginBottom: verticalSpacing,
        marginLeft: horizontalSpacing,
        marginRight: horizontalSpacing,
        marginTop: verticalSpacing,
        as: "hr",
        borderWidth: "borderWidth0",
        borderColor: "colorBorderWeak",
        borderStyle: "solid",
        borderBottomWidth: orientation === "horizontal" ? "borderWidth10" : null,
        borderLeftWidth: orientation === "vertical" ? "borderWidth10" : null,
        width: orientation === "horizontal" ? "auto" : null,
        height: orientation === "vertical" ? "auto" : null,
        ref
      })
    );
  }
);
Separator.displayName = "Separator";
Separator.propTypes = {
  element: import_prop_types.default.string,
  orientation: import_prop_types.default.oneOf(["horizontal", "vertical"]).isRequired,
  horizontalSpacing: import_style_props.isMarginTokenProp,
  verticalSpacing: import_style_props.isMarginTokenProp
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Separator
});
