// @twilio-paste/switch/dist/index.debug.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  Switch: () => Switch,
  SwitchContainer: () => SwitchContainer
});
module.exports = __toCommonJS(src_exports);

// src/SwitchContainer.tsx
var React = __toESM(require("react"));
var PropTypes = __toESM(require("prop-types"));
var import_text = require("@twilio-paste/text");
var import_uid_library = require("@twilio-paste/uid-library");
var import_label = require("@twilio-paste/label");
var import_media_object = require("@twilio-paste/media-object");
var import_help_text = require("@twilio-paste/help-text");
var SwitchContainer = React.forwardRef((_a, ref) => {
  var _b = _a, { element = "SWITCH_CONTAINER", children, disabled, helpText, label, id, required = false } = _b, props = __objRest(_b, ["element", "children", "disabled", "helpText", "label", "id", "required"]);
  const switchId = id ? id : (0, import_uid_library.useUID)();
  const helpTextId = (0, import_uid_library.useUID)();
  const labelId = (0, import_uid_library.useUID)();
  return /* @__PURE__ */ React.createElement(import_media_object.MediaObject, __spreadProps(__spreadValues({}, props), {
    element,
    ref
  }), /* @__PURE__ */ React.createElement(import_media_object.MediaFigure, {
    spacing: "space30"
  }, React.cloneElement(children, {
    id: switchId,
    labelId,
    helpTextId,
    disabled
  })), /* @__PURE__ */ React.createElement(import_media_object.MediaBody, null, /* @__PURE__ */ React.createElement(import_label.Label, {
    as: "div",
    htmlFor: switchId,
    marginBottom: "space0",
    disabled,
    ref,
    id: labelId,
    element: `${element}_LABEL`,
    required
  }, /* @__PURE__ */ React.createElement(import_text.Text, {
    as: "span",
    color: "currentColor",
    fontWeight: "fontWeightMedium"
  }, label != null ? label : /* @__PURE__ */ React.createElement(React.Fragment, null))), helpText ? /* @__PURE__ */ React.createElement(import_help_text.HelpText, {
    element: `${element}_HELP_TEXT`,
    marginTop: "space0",
    id: helpTextId
  }, helpText) : null));
});
SwitchContainer.displayName = "SwitchContainer";
SwitchContainer.propTypes = {
  children: PropTypes.node.isRequired,
  id: PropTypes.string,
  disabled: PropTypes.bool,
  element: PropTypes.string,
  helpText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  required: PropTypes.bool
};

// src/Switch.tsx
var React3 = __toESM(require("react"));
var PropTypes3 = __toESM(require("prop-types"));

// src/constants.ts
var SWITCH_WIDTH = "2.375rem";
var SWITCH_HEIGHT = "1rem";

// src/SwitchKnob.tsx
var React2 = __toESM(require("react"));
var PropTypes2 = __toESM(require("prop-types"));
var import_box = require("@twilio-paste/box");
var import_CheckboxCheckIcon = require("@twilio-paste/icons/cjs/CheckboxCheckIcon");
var import_theme = require("@twilio-paste/theme");
var getSwitchKnobTransformValue = (disabled, switchIsOn, isHovering) => {
  if (!disabled) {
    if (!switchIsOn && isHovering)
      return "translateX(5%)";
    if (switchIsOn && !isHovering)
      return "translateX(100%)";
    if (switchIsOn && isHovering)
      return "translateX(95%)";
  } else if (disabled) {
    if (!switchIsOn)
      return "translate(0%)";
    if (switchIsOn)
      return "translateX(100%)";
  }
  return "translate(0%)";
};
var SwitchKnob = React2.forwardRef((_a, ref) => {
  var _b = _a, { element, disabled = false, switchIsOn, isHovering, height } = _b, props = __objRest(_b, ["element", "disabled", "switchIsOn", "isHovering", "height"]);
  const theme = (0, import_theme.useTheme)();
  const SWITCH_PADDING = theme.space ? theme.space.space20 : "4px";
  const transformValue = React2.useMemo(() => getSwitchKnobTransformValue(disabled, switchIsOn, isHovering), [disabled, switchIsOn, isHovering]);
  return /* @__PURE__ */ React2.createElement(import_box.Box, {
    width: `calc(100% - ${SWITCH_PADDING})`,
    height: "100%",
    position: "absolute",
    transform: transformValue,
    transition: "transform .2s ease-in-out"
  }, /* @__PURE__ */ React2.createElement(import_box.Box, __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
    "aria-disabled": disabled,
    element: `${element}_KNOB`,
    ref,
    height,
    width: height,
    transform: switchIsOn ? "translateX(-100%)" : "translateX(0%)",
    transition: "transform .2s ease-in-out",
    borderColor: "colorBorder",
    borderWidth: "borderWidth10",
    borderRadius: "borderRadiusCircle",
    backgroundColor: "colorBackgroundBody",
    boxShadow: "shadowLow",
    _disabled: {
      boxShadow: "none"
    }
  }), /* @__PURE__ */ React2.createElement(import_box.Box, {
    display: "flex",
    height: "100%",
    alignItems: "center",
    justifyContent: "center",
    opacity: switchIsOn ? "1" : "0",
    transition: "opacity .2s ease-in-out, color .2s ease-in-out"
  }, /* @__PURE__ */ React2.createElement(import_CheckboxCheckIcon.CheckboxCheckIcon, {
    decorative: true,
    size: "sizeIcon05",
    element: `${element}_ICON`
  }))));
});
SwitchKnob.displayName = "SwitchKnob";
SwitchKnob.propTypes = {
  disabled: PropTypes2.bool,
  element: PropTypes2.string,
  switchIsOn: PropTypes2.bool.isRequired,
  isHovering: PropTypes2.bool.isRequired,
  height: PropTypes2.string.isRequired
};

// src/Switch.tsx
var import_box2 = require("@twilio-paste/box");
var styles = {
  backgroundColor: "colorBackgroundStronger",
  _checked_hover: {
    backgroundColor: "colorBackgroundPrimary",
    color: "colorTextLink"
  },
  _checked: {
    backgroundColor: "colorBackgroundPrimaryStronger",
    color: "colorTextLinkStronger"
  },
  _hover: {
    backgroundColor: "colorBackgroundStrongest",
    cursor: "pointer"
  },
  _disabled: {
    backgroundColor: "colorBackgroundStrong",
    color: "colorTextIcon",
    cursor: "not-allowed"
  },
  _focus: { boxShadow: "shadowFocus" }
};
var Switch = React3.forwardRef((_a, ref) => {
  var _b = _a, { element = "SWITCH", id, labelId, helpTextId, disabled = false, on = false, onClick } = _b, props = __objRest(_b, ["element", "id", "labelId", "helpTextId", "disabled", "on", "onClick"]);
  const [switchIsOn, setSwitchIsOn] = React3.useState(on);
  const [isHovering, setIsHovering] = React3.useState(false);
  const handleClick = React3.useCallback(() => {
    if (!disabled) {
      setSwitchIsOn(!switchIsOn);
      if (onClick)
        onClick();
    }
  }, [onClick, disabled, switchIsOn]);
  const handleKeyDown = React3.useCallback((event) => {
    if (event.key === " " || event.key === "Enter")
      setSwitchIsOn((prev) => !prev);
  }, []);
  return /* @__PURE__ */ React3.createElement(import_box2.Box, __spreadProps(__spreadValues(__spreadValues({}, (0, import_box2.safelySpreadBoxProps)(props)), styles), {
    as: "div",
    role: "switch",
    "aria-checked": switchIsOn,
    "aria-disabled": disabled,
    "aria-labelledby": labelId,
    "aria-describedby": helpTextId,
    element,
    id,
    ref,
    tabIndex: 0,
    outline: "none",
    position: "relative",
    display: "inline-block",
    boxSizing: "content-box",
    height: SWITCH_HEIGHT,
    width: SWITCH_WIDTH,
    overflow: "hidden",
    padding: "space10",
    borderColor: "colorBorder",
    borderWidth: "borderWidth10",
    borderRadius: "borderRadiusPill",
    transition: "background-color .2s ease-in-out, box-shadow .2s ease-in-out",
    onClick: handleClick,
    onKeyDown: handleKeyDown,
    onMouseEnter: () => {
      setIsHovering(true);
    },
    onMouseLeave: () => {
      setIsHovering(false);
    }
  }), /* @__PURE__ */ React3.createElement(SwitchKnob, {
    element,
    disabled,
    switchIsOn,
    isHovering,
    height: SWITCH_HEIGHT
  }));
});
Switch.displayName = "Switch";
Switch.propTypes = {
  disabled: PropTypes3.bool,
  element: PropTypes3.string,
  labelId: PropTypes3.string,
  id: PropTypes3.string,
  on: PropTypes3.bool,
  onClick: PropTypes3.func
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Switch,
  SwitchContainer
});
