// @twilio-paste/label/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  Label: () => Label,
  RequiredDot: () => RequiredDot
});
module.exports = __toCommonJS(src_exports);

// src/Label.tsx
var React2 = __toESM(require("react"));
var import_prop_types2 = __toESM(require("prop-types"));
var import_box2 = require("@twilio-paste/box");
var import_media_object = require("@twilio-paste/media-object");

// src/RequiredDot.tsx
var React = __toESM(require("react"));
var import_prop_types = __toESM(require("prop-types"));
var import_box = require("@twilio-paste/box");
var import_screen_reader_only = require("@twilio-paste/screen-reader-only");
var RequiredDot = (_a) => {
  var _b = _a, { element = "REQUIRED_DOT", i18nLabel = "" } = _b, props = __objRest(_b, ["element", "i18nLabel"]);
  return /* @__PURE__ */ React.createElement(
    import_box.Box,
    __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
      alignItems: "center",
      justifyContent: "center",
      as: "span",
      cursor: "pointer",
      display: "flex",
      height: "sizeIcon30",
      element: `${element}_WRAPPER`
    }),
    /* @__PURE__ */ React.createElement(
      import_box.Box,
      {
        as: "span",
        backgroundColor: "colorBackgroundRequired",
        borderRadius: "borderRadiusCircle",
        display: "block",
        height: "4px",
        width: "4px",
        element
      },
      /* @__PURE__ */ React.createElement(import_screen_reader_only.ScreenReaderOnly, null, i18nLabel)
    )
  );
};
RequiredDot.displayName = "RequiredDot";
RequiredDot.propTypes = {
  element: import_prop_types.default.string
};

// src/Label.tsx
var Label = React2.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      as = "label",
      marginBottom,
      required,
      disabled,
      children,
      variant,
      element = "LABEL",
      i18nRequiredLabel = ""
    } = _b, props = __objRest(_b, [
      "as",
      "marginBottom",
      "required",
      "disabled",
      "children",
      "variant",
      "element",
      "i18nRequiredLabel"
    ]);
    let textColor = "colorText";
    if (disabled && variant === "inverse") {
      textColor = "colorTextInverseWeak";
    } else if (disabled) {
      textColor = "colorTextWeak";
    } else if (variant === "inverse") {
      textColor = "colorTextInverse";
    }
    let cursor = "pointer";
    if (disabled) {
      cursor = "not-allowed";
    } else if (as === "legend" || as === "div") {
      cursor = "default";
    }
    return /* @__PURE__ */ React2.createElement(
      import_box2.Box,
      __spreadProps(__spreadValues({}, (0, import_box2.safelySpreadBoxProps)(props)), {
        as,
        borderBottomWidth: "borderWidth0",
        display: "block",
        marginBottom: marginBottom || "space20",
        paddingLeft: "space0",
        paddingRight: "space0",
        textTransform: "none",
        element,
        fontSize: "fontSize30",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight30",
        color: textColor,
        cursor,
        ref
      }),
      /* @__PURE__ */ React2.createElement(import_media_object.MediaObject, { verticalAlign: "top" }, required && /* @__PURE__ */ React2.createElement(import_media_object.MediaFigure, { spacing: "space20" }, /* @__PURE__ */ React2.createElement(RequiredDot, { element: `${element}_REQUIRED_DOT`, i18nLabel: i18nRequiredLabel })), /* @__PURE__ */ React2.createElement(import_media_object.MediaBody, null, children))
    );
  }
);
Label.displayName = "Label";
Label.propTypes = {
  as: import_prop_types2.default.oneOf(["label", "legend", "div"]),
  disabled: import_prop_types2.default.bool,
  element: import_prop_types2.default.string,
  htmlFor: import_prop_types2.default.string,
  marginBottom: import_prop_types2.default.oneOf(["space0"]),
  required: import_prop_types2.default.bool,
  i18nRequiredLabel: import_prop_types2.default.string,
  variant: import_prop_types2.default.oneOf(["default", "inverse"])
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Label,
  RequiredDot
});
