// twilio-sync/dist/itemsCache.js
/*
@license
The following license applies to all parts of this software except as
documented below.

    Copyright (c) 2016, Twilio, inc.
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are
    met:

      1. Redistributions of source code must retain the above copyright
         notice, this list of conditions and the following disclaimer.

      2. Redistributions in binary form must reproduce the above copyright
         notice, this list of conditions and the following disclaimer in
         the documentation and/or other materials provided with the
         distribution.

      3. Neither the name of Twilio nor the names of its contributors may
         be used to endorse or promote products derived from this software
         without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
    A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

This software includes loglevel under the following license.

    Copyright (c) 2013 Tim Perry

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the "Software"), to deal in the Software without
    restriction, including without limitation the rights to use,
    copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the
    Software is furnished to do so, subject to the following
    conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
    OTHER DEALINGS IN THE SOFTWARE.

This software includes Backoff library under the following license

    Copyright (C) 2012 Mathieu Turcotte

    Permission is hereby granted, free of charge, to any person obtaining a copy of
    this software and associated documentation files (the "Software"), to deal in
    the Software without restriction, including without limitation the rights to
    use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
    of the Software, and to permit persons to whom the Software is furnished to do
    so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.

This software includes Event-to-Promise library under the following license

    Copyright (c) 2014, Julien Fontanet <julien.fontanet@isonoe.net>.

    Permission to use, copy, modify, and/or distribute this software for any purpose
    with or without fee is hereby granted, provided that the above copyright notice
    and this permission notice appear in all copies.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
    FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
    OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
    TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
    THIS SOFTWARE.

*/
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tree = require('./utils/tree.js');

class Entry {
    constructor(value, lastEventId, isLeft = false, isRight = false) {
        this._data = value;
        this.lastEventId = (lastEventId || 0);
        this.isLeftBound = isLeft;
        this.isRightBound = isRight;
    }
    get isValid() {
        return true;
    }
    get value() {
        return this._data;
    }
}
class Tombstone {
    constructor(lastEventId) {
        this.lastEventId = lastEventId;
        this.isLeftBound = false;
        this.isRightBound = false;
    }
    get isValid() {
        return false;
    }
    get value() {
        return null;
    }
}
/// A cache for ranges of items in a collection.
/// Keeps track of the left and right bounds of the entire collection.
/// Keeps track of the left and right bounds of the pages of items.
class ItemsCache {
    constructor() {
        this.items = new tree.TreeMap();
        this.beginKey = null;
        this.endKey = null;
        this.items.clear(); // Set count to 0 instead of null.
    }
    /*
     * Used in tests.
     * @internal
     */
    get size() {
        return this.items.size;
    }
    get collectionBegin() {
        return this.beginKey;
    }
    get collectionEnd() {
        return this.endKey;
    }
    has(key) {
        var _a;
        let cacheEntry = this.items.get(key);
        return (_a = cacheEntry === null || cacheEntry === void 0 ? void 0 : cacheEntry.isValid) !== null && _a !== void 0 ? _a : false;
    }
    // Whether an entry exists in the cache either as a value or a tombstone.
    exists(key) {
        let cacheEntry = this.items.get(key);
        return !!cacheEntry;
    }
    // Return only cached value or null if it is missing or marked with a tombstone.
    getValue(key) {
        var _a;
        let cacheEntry = this.items.get(key);
        return (_a = cacheEntry === null || cacheEntry === void 0 ? void 0 : cacheEntry.value) !== null && _a !== void 0 ? _a : null;
    }
    // Return a cache entry, can either be missing (null), valid or a tombstone.
    get(key) {
        return this.items.get(key);
    }
    // Get an item next to the given key in specified search order.
    getNextItem(key, order, inclusive) {
        if (key === null) {
            return this.getFirstItem(order);
        }
        if (inclusive) {
            return this.getValue(key);
        }
        return this.getAdjacentItem(key, order);
    }
    // Find the first item in the cache based on query order.
    // @internal
    getFirstItem(order) {
        const key = (order === 'asc') ? this.beginKey : this.endKey;
        if (this.exists(key)) {
            const entry = this.items.get(key);
            if (entry.isValid) {
                return entry.value;
            }
            return this.getAdjacentItem(key, order);
        }
        return null;
    }
    // Find the first valid item next to the given one.
    // @internal
    getAdjacentItem(key, order) {
        const iterator = (order === 'asc') ? this.items.getIterator(key) : this.items.getReverseIterator(key);
        for (const [k, v] of iterator) {
            if ((order === 'asc' ? v.isRightBound : v.isLeftBound)) {
                break;
            }
            if (v.isValid) {
                return v.value;
            }
        }
        return null;
    }
    store(key, data, lastEventId, isLeft = true, isRight = true) {
        let cacheEntry = this.items.get(key);
        if (cacheEntry && cacheEntry.lastEventId > lastEventId) {
            return cacheEntry.value;
        }
        this.items.set(key, new Entry(data, lastEventId, cacheEntry ? cacheEntry.isLeftBound && isLeft : isLeft, cacheEntry ? cacheEntry.isRightBound && isRight : isRight));
        // Update the `begin` and `end` markers. - FIXME can't do it here, because we don't know the pagination limits
        if (this.beginKey && key < this.beginKey) {
            // this.items[begin].isLeftBound = False // FIXME continuity?
            this.beginKey = key;
        }
        if (this.endKey && this.endKey < key) {
            // cache.items[end].isRightBound = False
            this.endKey = key;
        }
        return data;
    }
    // Populate a contiguous range of items from a paginator page.
    populate(entries, isBegin, isEnd) {
        if (!entries.length) {
            return;
        }
        const newEntries = [...entries];
        // Backend-compatible lexicographical key sort
        newEntries.sort((x1, x2) => {
            if (x1.cacheKey === x2.cacheKey) {
                return 0;
            }
            if (x1.cacheKey > x2.cacheKey) {
                return 1;
            }
            return -1;
        });
        newEntries.forEach(((item, index) => {
            this.store(item.cacheKey, item, item.lastEventId, index === 0, index === newEntries.length - 1);
        }));
        if (isBegin) {
            const firstKey = newEntries[0].cacheKey;
            if (!this.beginKey || firstKey < this.beginKey) {
                this.beginKey = firstKey;
            }
        }
        if (isEnd) {
            const lastKey = newEntries[newEntries.length - 1].cacheKey;
            if (!this.endKey || this.endKey < lastKey) {
                this.endKey = lastKey;
            }
        }
    }
    delete(key, lastEventId) {
        let cacheEntry = this.items.get(key);
        /* Force delete when revision is unknown */
        if (cacheEntry === null || lastEventId === null || cacheEntry.lastEventId < lastEventId) {
            this.items.set(key, new Tombstone(lastEventId));
        }
    }
    // Shall we ignore the item with this key in an update?
    isKnown(key, lastEventId) {
        let cacheEntry = this.items.get(key);
        return cacheEntry && cacheEntry.lastEventId >= lastEventId;
    }
    *getIterator(key) {
        if (key === null && this.beginKey !== null) {
            key = this.beginKey;
        }
        if (key === null) {
            return;
        }
        for (let [k, v] of this.items.getIterator(key)) {
            if (!v.isValid) {
                continue;
            }
            yield v;
            if (v.isRightBound) {
                break;
            }
        }
    }
    *getReverseIterator(key) {
        if (key === null && this.endKey !== null) {
            key = this.endKey;
        }
        if (key === null) {
            return;
        }
        for (let [k, v] of this.items.getReverseIterator(key)) {
            if (!v.isValid) {
                continue;
            }
            yield v;
            if (v.isLeftBound) {
                break;
            }
        }
    }
    forEach(callbackfn) {
        if (this.items) {
            for (let [key, entry] of this.items) {
                if (entry.isValid) {
                    callbackfn(key, entry.value);
                }
            }
        }
    }
}

exports.ItemsCache = ItemsCache;
exports["default"] = ItemsCache;
