// @twilio-paste/help-text/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  HelpText: () => HelpText,
  HelpTextVariants: () => HelpTextVariants
});
module.exports = __toCommonJS(src_exports);

// src/HelpText.tsx
var React = __toESM(require("react"));
var import_prop_types = __toESM(require("prop-types"));
var import_box = require("@twilio-paste/box");
var import_ErrorIcon = require("@twilio-paste/icons/cjs/ErrorIcon");
var import_SuccessIcon = require("@twilio-paste/icons/cjs/SuccessIcon");
var import_WarningIcon = require("@twilio-paste/icons/cjs/WarningIcon");
var HelpTextVariants = {
  DEFAULT: "default",
  ERROR: "error",
  ERROR_INVERSE: "error_inverse",
  INVERSE: "inverse",
  SUCCESS: "success",
  WARNING: "warning"
};
var VariantOptions = {
  [HelpTextVariants.DEFAULT]: {
    textColor: "colorTextWeak",
    icon: null
  },
  [HelpTextVariants.INVERSE]: {
    textColor: "colorTextInverseWeak",
    icon: null
  },
  [HelpTextVariants.ERROR]: {
    textColor: "colorTextError",
    icon: /* @__PURE__ */ React.createElement(import_box.Box, { flexShrink: 0 }, /* @__PURE__ */ React.createElement(import_ErrorIcon.ErrorIcon, { color: "colorTextError", decorative: true, size: "sizeIcon20" }))
  },
  [HelpTextVariants.ERROR_INVERSE]: {
    textColor: "colorTextErrorWeak",
    icon: /* @__PURE__ */ React.createElement(import_box.Box, { flexShrink: 0 }, /* @__PURE__ */ React.createElement(import_ErrorIcon.ErrorIcon, { color: "colorTextErrorWeak", decorative: true, size: "sizeIcon20" }))
  },
  [HelpTextVariants.SUCCESS]: {
    textColor: "colorTextSuccess",
    icon: /* @__PURE__ */ React.createElement(import_box.Box, { flexShrink: 0 }, /* @__PURE__ */ React.createElement(import_SuccessIcon.SuccessIcon, { color: "colorTextIconSuccess", decorative: true, size: "sizeIcon20" }))
  },
  [HelpTextVariants.WARNING]: {
    textColor: "colorTextWarning",
    icon: /* @__PURE__ */ React.createElement(import_box.Box, { flexShrink: 0 }, /* @__PURE__ */ React.createElement(import_WarningIcon.WarningIcon, { color: "colorTextWarning", decorative: true, size: "sizeIcon20" }))
  }
};
var HelpText = React.forwardRef(
  (_a, ref) => {
    var _b = _a, { marginTop, children, variant = "default", element = "HELP_TEXT" } = _b, props = __objRest(_b, ["marginTop", "children", "variant", "element"]);
    const { textColor, icon } = VariantOptions[variant];
    return /* @__PURE__ */ React.createElement(
      import_box.Box,
      __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
        display: "flex",
        columnGap: "space20",
        marginTop: marginTop || "space30",
        ref,
        element,
        variant,
        as: "div",
        color: textColor,
        fontSize: "fontSize30",
        lineHeight: "lineHeight30"
      }),
      icon,
      /* @__PURE__ */ React.createElement("span", null, children)
    );
  }
);
HelpText.displayName = "HelpText";
HelpText.propTypes = {
  marginTop: import_prop_types.default.oneOf(["space0"]),
  element: import_prop_types.default.string
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  HelpText,
  HelpTextVariants
});
