// @twilio-paste/heading/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  Heading: () => Heading,
  HeadingPropTypes: () => HeadingPropTypes
});
module.exports = __toCommonJS(src_exports);
var React = __toESM(require("react"));
var import_text = require("@twilio-paste/text");

// src/PropTypes.ts
var import_prop_types = __toESM(require("prop-types"));
var HeadingPropTypes = {
  as: import_prop_types.default.oneOf(["h1", "h2", "h3", "h4", "h5", "h6", "div", "label", "span"]).isRequired,
  element: import_prop_types.default.string,
  marginBottom: import_prop_types.default.oneOf(["space0"]),
  variant: import_prop_types.default.oneOf([
    "heading10",
    "heading20",
    "heading30",
    "heading40",
    "heading50",
    "heading60"
  ]).isRequired,
  display: import_prop_types.default.string
};

// src/index.tsx
function getHeadingProps(headingVariant, marginBottom) {
  switch (headingVariant) {
    case "heading10":
      return {
        marginBottom: marginBottom || "space70",
        fontSize: "fontSize90",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight90"
      };
    case "heading30":
      return {
        marginBottom: marginBottom || "space50",
        fontSize: "fontSize60",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight60"
      };
    case "heading40":
      return {
        marginBottom: marginBottom || "space40",
        fontSize: "fontSize40",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight40"
      };
    case "heading50":
      return {
        marginBottom: marginBottom || "space30",
        fontSize: "fontSize30",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight30"
      };
    case "heading60":
      return {
        marginBottom: marginBottom || "space30",
        fontSize: "fontSize20",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight20"
      };
    case "heading20":
    default:
      return {
        marginBottom: marginBottom || "space60",
        fontSize: "fontSize70",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight70"
      };
  }
}
var Heading = React.forwardRef(
  (_a, ref) => {
    var _b = _a, { as, children, display = "block", element = "HEADING", id, marginBottom, variant } = _b, props = __objRest(_b, ["as", "children", "display", "element", "id", "marginBottom", "variant"]);
    return /* @__PURE__ */ React.createElement(
      import_text.Text,
      __spreadProps(__spreadValues(__spreadValues({}, (0, import_text.safelySpreadTextProps)(props)), getHeadingProps(variant, marginBottom)), {
        as,
        display,
        element,
        id,
        color: "colorText",
        variant,
        ref
      }),
      children
    );
  }
);
Heading.displayName = "Heading";
Heading.propTypes = HeadingPropTypes;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Heading,
  HeadingPropTypes
});
