// @twilio-paste/alert/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  Alert: () => Alert,
  AlertBackgroundColors: () => AlertBackgroundColors,
  AlertBorderColors: () => AlertBorderColors,
  AlertRoles: () => AlertRoles,
  AlertVariants: () => AlertVariants
});
module.exports = __toCommonJS(src_exports);
var React = __toESM(require("react"));
var import_prop_types = __toESM(require("prop-types"));
var import_box = require("@twilio-paste/box");
var import_media_object = require("@twilio-paste/media-object");
var import_button = require("@twilio-paste/button");
var import_screen_reader_only = require("@twilio-paste/screen-reader-only");
var import_CloseIcon = require("@twilio-paste/icons/cjs/CloseIcon");
var import_ErrorIcon = require("@twilio-paste/icons/cjs/ErrorIcon");
var import_NeutralIcon = require("@twilio-paste/icons/cjs/NeutralIcon");
var import_WarningIcon = require("@twilio-paste/icons/cjs/WarningIcon");
var AlertRoles = {
  ERROR: "alert",
  NEUTRAL: "status",
  WARNING: "alert"
};
var AlertVariants = {
  ERROR: "error",
  NEUTRAL: "neutral",
  WARNING: "warning"
};
var AlertBackgroundColors = {
  ERROR: "colorBackgroundErrorWeakest",
  NEUTRAL: "colorBackgroundNeutralWeakest",
  WARNING: "colorBackgroundWarningWeakest"
};
var AlertBorderColors = {
  ERROR: "colorBorderErrorWeak",
  NEUTRAL: "colorBorderNeutralWeak",
  WARNING: "colorBorderWarningWeak"
};
var renderAlertIcon = (variant, element, title) => {
  switch (variant) {
    case AlertVariants.ERROR:
      return /* @__PURE__ */ React.createElement(
        import_ErrorIcon.ErrorIcon,
        {
          element: `${element}_ICON`,
          color: "colorTextIconError",
          decorative: false,
          title,
          size: "sizeIcon20"
        }
      );
    case AlertVariants.WARNING:
      return /* @__PURE__ */ React.createElement(
        import_WarningIcon.WarningIcon,
        {
          element: `${element}_ICON`,
          color: "colorTextIconWarning",
          decorative: false,
          title,
          size: "sizeIcon20"
        }
      );
    case AlertVariants.NEUTRAL:
    default:
      return /* @__PURE__ */ React.createElement(
        import_NeutralIcon.NeutralIcon,
        {
          element: `${element}_ICON`,
          color: "colorTextIconNeutral",
          decorative: false,
          title,
          size: "sizeIcon20"
        }
      );
  }
};
var Alert = React.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      children,
      onDismiss,
      variant,
      role,
      element = "ALERT",
      i18nDismissLabel = "Dismiss alert",
      i18nErrorLabel = "(error)",
      i18nNeutralLabel = "(information)",
      i18nWarningLabel = "(warning)"
    } = _b, props = __objRest(_b, [
      "children",
      "onDismiss",
      "variant",
      "role",
      "element",
      "i18nDismissLabel",
      "i18nErrorLabel",
      "i18nNeutralLabel",
      "i18nWarningLabel"
    ]);
    const i18nLabelVariantMap = {
      error: i18nErrorLabel,
      neutral: i18nNeutralLabel,
      warning: i18nWarningLabel
    };
    return /* @__PURE__ */ React.createElement(
      import_box.Box,
      __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
        backgroundColor: AlertBackgroundColors[variant.toUpperCase()],
        borderColor: AlertBorderColors[variant.toUpperCase()],
        borderBottomWidth: "borderWidth20",
        borderBottomStyle: "solid",
        paddingLeft: "space60",
        paddingRight: "space60",
        paddingTop: "space50",
        paddingBottom: "space50",
        element,
        variant,
        ref,
        role: role != null ? role : AlertRoles[variant.toUpperCase()]
      }),
      /* @__PURE__ */ React.createElement(import_media_object.MediaObject, { as: "div" }, /* @__PURE__ */ React.createElement(import_media_object.MediaFigure, { as: "div", spacing: "space30" }, renderAlertIcon(variant, element, i18nLabelVariantMap[variant])), /* @__PURE__ */ React.createElement(import_media_object.MediaBody, { as: "div" }, children), onDismiss && typeof onDismiss === "function" && /* @__PURE__ */ React.createElement(import_media_object.MediaFigure, { align: "end", spacing: "space60" }, /* @__PURE__ */ React.createElement(import_button.Button, { onClick: onDismiss, variant: "secondary_icon", size: "reset", element: `${element}_DISMISS_BUTTON` }, /* @__PURE__ */ React.createElement(import_CloseIcon.CloseIcon, { element: `${element}_DISMISS_ICON`, decorative: true, size: "sizeIcon20" }), /* @__PURE__ */ React.createElement(import_screen_reader_only.ScreenReaderOnly, null, i18nDismissLabel))))
    );
  }
);
Alert.displayName = "Alert";
Alert.propTypes = {
  children: import_prop_types.default.node.isRequired,
  onDismiss: import_prop_types.default.func,
  role: import_prop_types.default.string,
  variant: import_prop_types.default.oneOf(Object.values(AlertVariants)).isRequired,
  element: import_prop_types.default.string,
  i18nDismissLabel: import_prop_types.default.string,
  i18nErrorLabel: import_prop_types.default.string,
  i18nNeutralLabel: import_prop_types.default.string,
  i18nWarningLabel: import_prop_types.default.string
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Alert,
  AlertBackgroundColors,
  AlertBorderColors,
  AlertRoles,
  AlertVariants
});
