// @twilio-paste/popover/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  Popover: () => Popover,
  PopoverBadgeButton: () => PopoverBadgeButton,
  PopoverButton: () => PopoverButton,
  PopoverContainer: () => PopoverContainer,
  usePopoverState: () => import_non_modal_dialog_primitive.useNonModalDialogPrimitiveState
});
module.exports = __toCommonJS(src_exports);

// src/PopoverContainer.tsx
var React2 = __toESM(require("react"));
var import_non_modal_dialog_primitive = require("@twilio-paste/non-modal-dialog-primitive");

// src/PopoverContext.tsx
var React = __toESM(require("react"));
var PopoverContext = React.createContext({});

// src/PopoverContainer.tsx
var BasePopoverContainer = (_a) => {
  var _b = _a, {
    children,
    gutter,
    placement,
    state
  } = _b, initialState = __objRest(_b, [
    "children",
    "gutter",
    "placement",
    "state"
  ]);
  const popover = state || (0, import_non_modal_dialog_primitive.useNonModalDialogPrimitiveState)(__spreadValues({ gutter, modal: true, placement }, initialState));
  return /* @__PURE__ */ React2.createElement(PopoverContext.Provider, { value: popover }, children);
};
BasePopoverContainer.displayName = "BasePopoverContainer";
var PopoverContainer = React2.memo(BasePopoverContainer);
PopoverContainer.displayName = "PopoverContainer";

// src/Popover.tsx
var React5 = __toESM(require("react"));
var import_prop_types2 = __toESM(require("prop-types"));
var import_box2 = require("@twilio-paste/box");
var import_button = require("@twilio-paste/button");
var import_CloseIcon = require("@twilio-paste/icons/cjs/CloseIcon");
var import_theme2 = require("@twilio-paste/theme");
var import_non_modal_dialog_primitive3 = require("@twilio-paste/non-modal-dialog-primitive");

// ../screen-reader-only/dist/index.es.js
var React3 = __toESM(require("react"));
var import_prop_types = require("prop-types");
var import_box = require("@twilio-paste/box");
var __defProp2 = Object.defineProperty;
var __defProps2 = Object.defineProperties;
var __getOwnPropDescs2 = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols2 = Object.getOwnPropertySymbols;
var __hasOwnProp2 = Object.prototype.hasOwnProperty;
var __propIsEnum2 = Object.prototype.propertyIsEnumerable;
var __defNormalProp2 = (obj, key, value) => key in obj ? __defProp2(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues2 = (a, b) => {
  for (var prop in b || (b = {}))
    __hasOwnProp2.call(b, prop) && __defNormalProp2(a, prop, b[prop]);
  if (__getOwnPropSymbols2)
    for (var prop of __getOwnPropSymbols2(b))
      __propIsEnum2.call(b, prop) && __defNormalProp2(a, prop, b[prop]);
  return a;
};
var __spreadProps2 = (a, b) => __defProps2(a, __getOwnPropDescs2(b));
var __objRest2 = (source, exclude) => {
  var target = {};
  for (var prop in source)
    __hasOwnProp2.call(source, prop) && exclude.indexOf(prop) < 0 && (target[prop] = source[prop]);
  if (source != null && __getOwnPropSymbols2)
    for (var prop of __getOwnPropSymbols2(source))
      exclude.indexOf(prop) < 0 && __propIsEnum2.call(source, prop) && (target[prop] = source[prop]);
  return target;
};
var ScreenReaderOnly = React3.forwardRef((_a, ref) => {
  var _b = _a, { as = "span", children } = _b, props = __objRest2(_b, ["as", "children"]);
  return React3.createElement(import_box.Box, __spreadProps2(__spreadValues2({}, props), { as, border: "none", clip: "rect(0 0 0 0)", height: "1px", margin: "spaceNegative10", overflow: "hidden", padding: "space0", position: "absolute", ref, textTransform: "none", whiteSpace: "nowrap", width: "1px" }), children);
});
ScreenReaderOnly.displayName = "ScreenReaderOnly";

// src/PopoverArrow.tsx
var React4 = __toESM(require("react"));
var import_non_modal_dialog_primitive2 = require("@twilio-paste/non-modal-dialog-primitive");
var import_theme = require("@twilio-paste/theme");
var PopoverArrow = (props) => {
  const theme = (0, import_theme.useTheme)();
  return /* @__PURE__ */ React4.createElement(
    import_non_modal_dialog_primitive2.NonModalDialogArrowPrimitive,
    __spreadProps(__spreadValues({}, props), {
      size: theme.fontSizes.fontSize70,
      stroke: theme.borderColors.colorBorderWeaker,
      fill: theme.backgroundColors.colorBackgroundBody
    })
  );
};
PopoverArrow.displayName = "PopoverArrow";

// src/Popover.tsx
var StyledPopover = React5.forwardRef((_a, ref) => {
  var _b = _a, { style } = _b, props = __objRest(_b, ["style"]);
  return /* @__PURE__ */ React5.createElement(
    import_box2.Box,
    __spreadProps(__spreadValues({}, (0, import_box2.safelySpreadBoxProps)(props)), {
      backgroundColor: "colorBackgroundBody",
      borderStyle: "solid",
      borderWidth: "borderWidth10",
      borderColor: "colorBorderWeaker",
      borderRadius: "borderRadius20",
      boxShadow: "shadowCard",
      maxWidth: "size50",
      zIndex: "zIndex80",
      _focus: { outline: "none" },
      style,
      ref
    })
  );
});
StyledPopover.displayName = "StyledPopover";
var Popover = React5.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "POPOVER", i18nDismissLabel = "Close popover" } = _b, props = __objRest(_b, ["children", "element", "i18nDismissLabel"]);
    const popover = React5.useContext(PopoverContext);
    return /* @__PURE__ */ React5.createElement(import_non_modal_dialog_primitive3.NonModalDialogPrimitive, __spreadProps(__spreadValues(__spreadValues({}, popover), props), { as: StyledPopover, ref, preventBodyScroll: false }), /* @__PURE__ */ React5.createElement(import_theme2.StyledBase, null, /* @__PURE__ */ React5.createElement(PopoverArrow, __spreadValues({}, popover)), /* @__PURE__ */ React5.createElement(import_box2.Box, { element, paddingX: "space80", paddingY: "space70" }, /* @__PURE__ */ React5.createElement(import_box2.Box, { position: "absolute", right: 8, top: 8 }, /* @__PURE__ */ React5.createElement(
      import_button.Button,
      {
        element: `${element}_CLOSE_BUTTON`,
        variant: "secondary_icon",
        size: "reset",
        onClick: popover.hide
      },
      /* @__PURE__ */ React5.createElement(import_CloseIcon.CloseIcon, { element: `${element}_CLOSE_ICON`, decorative: true, size: "sizeIcon10" }),
      /* @__PURE__ */ React5.createElement(ScreenReaderOnly, null, i18nDismissLabel)
    )), children)));
  }
);
Popover.propTypes = {
  "aria-label": import_prop_types2.default.string.isRequired,
  children: import_prop_types2.default.node.isRequired,
  element: import_prop_types2.default.string,
  i18nDismissLabel: import_prop_types2.default.string
};
Popover.displayName = "Popover";

// src/PopoverButton.tsx
var React6 = __toESM(require("react"));
var import_prop_types3 = __toESM(require("prop-types"));
var import_button2 = require("@twilio-paste/button");
var import_non_modal_dialog_primitive4 = require("@twilio-paste/non-modal-dialog-primitive");
var PopoverButton = React6.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "POPOVER_BUTTON" } = _b, popoverButtonProps = __objRest(_b, ["children", "element"]);
    const popover = React6.useContext(PopoverContext);
    return /* @__PURE__ */ React6.createElement(
      import_non_modal_dialog_primitive4.NonModalDialogDisclosurePrimitive,
      __spreadProps(__spreadValues(__spreadValues({
        element
      }, popover), popoverButtonProps), {
        as: import_button2.Button,
        ref
      }),
      children
    );
  }
);
PopoverButton.propTypes = {
  id: import_prop_types3.default.string,
  children: import_prop_types3.default.node.isRequired,
  toggle: import_prop_types3.default.func,
  element: import_prop_types3.default.string
};
PopoverButton.displayName = "PopoverButton";

// src/PopoverBadgeButton.tsx
var React7 = __toESM(require("react"));
var import_prop_types4 = __toESM(require("prop-types"));
var import_badge = require("@twilio-paste/badge");
var import_non_modal_dialog_primitive5 = require("@twilio-paste/non-modal-dialog-primitive");
var ButtonBadge = React7.forwardRef(function ButtonBadge2(_a, ref) {
  var _b = _a, { children } = _b, props = __objRest(_b, ["children"]);
  return /* @__PURE__ */ React7.createElement(import_badge.Badge, __spreadProps(__spreadValues({}, props), { as: "button", ref }), children);
});
ButtonBadge.displayName = "ButtonBadge";
var PopoverBadgeButton = React7.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "POPOVER_BADGE" } = _b, popoverButtonProps = __objRest(_b, ["children", "element"]);
    const popover = React7.useContext(PopoverContext);
    return /* @__PURE__ */ React7.createElement(
      import_non_modal_dialog_primitive5.NonModalDialogDisclosurePrimitive,
      __spreadProps(__spreadValues(__spreadValues({
        element
      }, popover), popoverButtonProps), {
        as: ButtonBadge,
        ref
      }),
      children
    );
  }
);
PopoverBadgeButton.propTypes = {
  id: import_prop_types4.default.string,
  toggle: import_prop_types4.default.func,
  element: import_prop_types4.default.string
};
PopoverBadgeButton.displayName = "PopoverBadgeButton";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Popover,
  PopoverBadgeButton,
  PopoverButton,
  PopoverContainer,
  usePopoverState
});
