// @twilio-paste/toast/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  AnimatedToast: () => AnimatedToast,
  Toast: () => Toast,
  ToastContainer: () => ToastContainer,
  Toaster: () => Toaster,
  useToaster: () => useToaster
});
module.exports = __toCommonJS(src_exports);

// src/Toast.tsx
var React6 = __toESM(require("react"));
var import_button = require("@twilio-paste/button");
var import_CloseIcon = require("@twilio-paste/icons/cjs/CloseIcon");
var import_ErrorIcon = require("@twilio-paste/icons/cjs/ErrorIcon");
var import_NeutralIcon = require("@twilio-paste/icons/cjs/NeutralIcon");
var import_SuccessIcon = require("@twilio-paste/icons/cjs/SuccessIcon");
var import_WarningIcon = require("@twilio-paste/icons/cjs/WarningIcon");
var import_media_object = require("@twilio-paste/media-object");

// ../screen-reader-only/dist/index.es.js
var React = __toESM(require("react"));
var import_prop_types = require("prop-types");
var import_box = require("@twilio-paste/box");
var __defProp2 = Object.defineProperty;
var __defProps2 = Object.defineProperties;
var __getOwnPropDescs2 = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols2 = Object.getOwnPropertySymbols;
var __hasOwnProp2 = Object.prototype.hasOwnProperty;
var __propIsEnum2 = Object.prototype.propertyIsEnumerable;
var __defNormalProp2 = (obj, key, value) => key in obj ? __defProp2(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues2 = (a, b) => {
  for (var prop in b || (b = {}))
    __hasOwnProp2.call(b, prop) && __defNormalProp2(a, prop, b[prop]);
  if (__getOwnPropSymbols2)
    for (var prop of __getOwnPropSymbols2(b))
      __propIsEnum2.call(b, prop) && __defNormalProp2(a, prop, b[prop]);
  return a;
};
var __spreadProps2 = (a, b) => __defProps2(a, __getOwnPropDescs2(b));
var __objRest2 = (source, exclude) => {
  var target = {};
  for (var prop in source)
    __hasOwnProp2.call(source, prop) && exclude.indexOf(prop) < 0 && (target[prop] = source[prop]);
  if (source != null && __getOwnPropSymbols2)
    for (var prop of __getOwnPropSymbols2(source))
      exclude.indexOf(prop) < 0 && __propIsEnum2.call(source, prop) && (target[prop] = source[prop]);
  return target;
};
var ScreenReaderOnly = React.forwardRef((_a, ref) => {
  var _b = _a, { as = "span", children } = _b, props = __objRest2(_b, ["as", "children"]);
  return React.createElement(import_box.Box, __spreadProps2(__spreadValues2({}, props), { as, border: "none", clip: "rect(0 0 0 0)", height: "1px", margin: "spaceNegative10", overflow: "hidden", padding: "space0", position: "absolute", ref, textTransform: "none", whiteSpace: "nowrap", width: "1px" }), children);
});
ScreenReaderOnly.displayName = "ScreenReaderOnly";

// src/ErrorToast.tsx
var React2 = __toESM(require("react"));
var import_box2 = require("@twilio-paste/box");

// src/propTypes.ts
var import_prop_types2 = __toESM(require("prop-types"));

// src/constants.ts
var ToastVariantObject = {
  ERROR: "error",
  NEUTRAL: "neutral",
  SUCCESS: "success",
  WARNING: "warning"
};

// src/propTypes.ts
var ToastPropTypes = {
  children: import_prop_types2.default.node.isRequired,
  onDismiss: import_prop_types2.default.func,
  variant: import_prop_types2.default.oneOf(["error", "neutral", "success", "warning"]),
  i18nDismissLabel: import_prop_types2.default.string,
  i18nErrorLabel: import_prop_types2.default.string,
  i18nNeutralLabel: import_prop_types2.default.string,
  i18nSuccessLabel: import_prop_types2.default.string,
  i18nWarningLabel: import_prop_types2.default.string
};
var ToastPortalPropTypes = {
  children: import_prop_types2.default.node.isRequired,
  __console_patch: import_prop_types2.default.bool
};
var ToastContainerPropTypes = {
  children: import_prop_types2.default.node.isRequired,
  __console_patch: import_prop_types2.default.bool
};
var ToasterPropTypes = {
  toasts: import_prop_types2.default.arrayOf(
    import_prop_types2.default.shape({
      id: import_prop_types2.default.string.isRequired,
      message: import_prop_types2.default.node.isRequired,
      setFocus: import_prop_types2.default.bool,
      variant: import_prop_types2.default.oneOf(Object.values(ToastVariantObject)).isRequired
    })
  ).isRequired,
  pop: import_prop_types2.default.func.isRequired,
  __console_patch: import_prop_types2.default.bool
};

// src/ErrorToast.tsx
var ErrorToast = React2.forwardRef((_a, ref) => {
  var _b = _a, { element = "TOAST", variant } = _b, props = __objRest(_b, ["element", "variant"]);
  return /* @__PURE__ */ React2.createElement(
    import_box2.Box,
    __spreadProps(__spreadValues({}, (0, import_box2.safelySpreadBoxProps)(props)), {
      ref,
      backgroundColor: "colorBackgroundBody",
      borderColor: "colorBorderWeaker",
      borderRadius: "borderRadius20",
      borderStyle: "solid",
      borderWidth: "borderWidth10",
      borderLeftWidth: "borderWidth30",
      borderLeftColor: "colorBorderErrorWeak",
      boxShadow: "shadowHigh",
      element,
      padding: "space60",
      variant,
      width: "size40"
    }),
    props.children
  );
});
ErrorToast.displayName = "ErrorToast";
ErrorToast.propTypes = ToastPropTypes;

// src/NeutralToast.tsx
var React3 = __toESM(require("react"));
var import_box3 = require("@twilio-paste/box");
var NeutralToast = React3.forwardRef((_a, ref) => {
  var _b = _a, { element = "TOAST", variant } = _b, props = __objRest(_b, ["element", "variant"]);
  return /* @__PURE__ */ React3.createElement(
    import_box3.Box,
    __spreadProps(__spreadValues({}, (0, import_box3.safelySpreadBoxProps)(props)), {
      ref,
      backgroundColor: "colorBackgroundBody",
      borderColor: "colorBorderWeaker",
      borderRadius: "borderRadius20",
      borderStyle: "solid",
      borderWidth: "borderWidth10",
      borderLeftWidth: "borderWidth30",
      borderLeftColor: "colorBorderNeutralWeak",
      boxShadow: "shadowHigh",
      element,
      padding: "space60",
      variant,
      width: "size40"
    }),
    props.children
  );
});
NeutralToast.displayName = "NeutralToast";
NeutralToast.propTypes = ToastPropTypes;

// src/SuccessToast.tsx
var React4 = __toESM(require("react"));
var import_box4 = require("@twilio-paste/box");
var SuccessToast = React4.forwardRef((_a, ref) => {
  var _b = _a, { element = "TOAST", variant } = _b, props = __objRest(_b, ["element", "variant"]);
  return /* @__PURE__ */ React4.createElement(
    import_box4.Box,
    __spreadProps(__spreadValues({}, (0, import_box4.safelySpreadBoxProps)(props)), {
      ref,
      backgroundColor: "colorBackgroundBody",
      borderColor: "colorBorderWeaker",
      borderRadius: "borderRadius20",
      borderStyle: "solid",
      borderWidth: "borderWidth10",
      borderLeftWidth: "borderWidth30",
      borderLeftColor: "colorBorderSuccessWeak",
      boxShadow: "shadowHigh",
      element,
      padding: "space60",
      variant,
      width: "size40"
    }),
    props.children
  );
});
SuccessToast.displayName = "SuccessToast";
SuccessToast.propTypes = ToastPropTypes;

// src/WarningToast.tsx
var React5 = __toESM(require("react"));
var import_box5 = require("@twilio-paste/box");
var WarningToast = React5.forwardRef((_a, ref) => {
  var _b = _a, { element = "TOAST", variant } = _b, props = __objRest(_b, ["element", "variant"]);
  return /* @__PURE__ */ React5.createElement(
    import_box5.Box,
    __spreadProps(__spreadValues({}, (0, import_box5.safelySpreadBoxProps)(props)), {
      ref,
      backgroundColor: "colorBackgroundBody",
      borderColor: "colorBorderWeaker",
      borderRadius: "borderRadius20",
      borderStyle: "solid",
      borderWidth: "borderWidth10",
      borderLeftWidth: "borderWidth30",
      borderLeftColor: "colorBorderWarningWeak",
      boxShadow: "shadowHigh",
      element,
      padding: "space60",
      variant,
      width: "size40"
    }),
    props.children
  );
});
WarningToast.displayName = "WarningToast";
WarningToast.propTypes = ToastPropTypes;

// src/Toast.tsx
var ToastComponentVariants = {
  error: ErrorToast,
  neutral: NeutralToast,
  success: SuccessToast,
  warning: WarningToast
};
var renderToastIcon = (variant, title, element) => {
  switch (variant) {
    case ToastVariantObject.ERROR:
      return /* @__PURE__ */ React6.createElement(
        import_ErrorIcon.ErrorIcon,
        {
          color: "colorTextIconError",
          decorative: false,
          title,
          element: `${element}_ICON`,
          size: "sizeIcon20"
        }
      );
    case ToastVariantObject.SUCCESS:
      return /* @__PURE__ */ React6.createElement(
        import_SuccessIcon.SuccessIcon,
        {
          color: "colorTextIconSuccess",
          decorative: false,
          title,
          element: `${element}_ICON`,
          size: "sizeIcon20"
        }
      );
    case ToastVariantObject.WARNING:
      return /* @__PURE__ */ React6.createElement(
        import_WarningIcon.WarningIcon,
        {
          color: "colorTextIconWarning",
          decorative: false,
          title,
          element: `${element}_ICON`,
          size: "sizeIcon20"
        }
      );
    case ToastVariantObject.NEUTRAL:
    default:
      return /* @__PURE__ */ React6.createElement(
        import_NeutralIcon.NeutralIcon,
        {
          color: "colorTextIconNeutral",
          decorative: false,
          title,
          element: `${element}_ICON`,
          size: "sizeIcon20"
        }
      );
  }
};
var Toast = React6.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      children,
      onDismiss,
      variant = "neutral",
      element = "TOAST",
      setFocus,
      i18nDismissLabel = "Dismiss toast",
      i18nErrorLabel = "(error)",
      i18nNeutralLabel = "(information)",
      i18nSuccessLabel = "(success)",
      i18nWarningLabel = "(warning)"
    } = _b, props = __objRest(_b, [
      "children",
      "onDismiss",
      "variant",
      "element",
      "setFocus",
      "i18nDismissLabel",
      "i18nErrorLabel",
      "i18nNeutralLabel",
      "i18nSuccessLabel",
      "i18nWarningLabel"
    ]);
    const ToastComponent = ToastComponentVariants[variant];
    const buttonRef = React6.useRef(null);
    React6.useEffect(() => {
      if (setFocus && buttonRef.current) {
        buttonRef.current.focus({ preventScroll: true });
      }
    }, [setFocus]);
    const i18nVariants = {
      error: i18nErrorLabel,
      neutral: i18nNeutralLabel,
      success: i18nSuccessLabel,
      warning: i18nWarningLabel
    };
    return /* @__PURE__ */ React6.createElement(ToastComponent, __spreadValues({ role: "status", variant, element, ref }, props), /* @__PURE__ */ React6.createElement(import_media_object.MediaObject, { as: "div" }, /* @__PURE__ */ React6.createElement(import_media_object.MediaFigure, { as: "div", spacing: "space60" }, renderToastIcon(variant, i18nVariants[variant], element)), /* @__PURE__ */ React6.createElement(import_media_object.MediaBody, { as: "div" }, children), onDismiss && typeof onDismiss === "function" && /* @__PURE__ */ React6.createElement(import_media_object.MediaFigure, { align: "end", spacing: "space40" }, /* @__PURE__ */ React6.createElement(
      import_button.Button,
      {
        onClick: onDismiss,
        variant: "secondary_icon",
        ref: buttonRef,
        size: "reset",
        element: `${element}_CLOSE_BUTTON`
      },
      /* @__PURE__ */ React6.createElement(import_CloseIcon.CloseIcon, { decorative: true, size: "sizeIcon20", element: `${element}_CLOSE_ICON` }),
      /* @__PURE__ */ React6.createElement(ScreenReaderOnly, null, i18nDismissLabel)
    ))));
  }
);
Toast.displayName = "Toast";
Toast.propTypes = ToastPropTypes;

// src/ToastContainer.tsx
var React8 = __toESM(require("react"));
var import_stack = require("@twilio-paste/stack");

// src/ToastPortal.tsx
var React7 = __toESM(require("react"));
var import_theme = require("@twilio-paste/theme");
var import_reakit_library = require("@twilio-paste/reakit-library");
var import_box6 = require("@twilio-paste/box");
var ToastPortal = React7.forwardRef((_a, ref) => {
  var _b = _a, { children, __console_patch } = _b, props = __objRest(_b, ["children", "__console_patch"]);
  return /* @__PURE__ */ React7.createElement(import_reakit_library.Portal, null, /* @__PURE__ */ React7.createElement(import_theme.StyledBase, null, /* @__PURE__ */ React7.createElement(
    import_box6.Box,
    __spreadProps(__spreadValues({}, (0, import_box6.safelySpreadBoxProps)(props)), {
      position: "fixed",
      right: "space40",
      top: "space90",
      zIndex: "zIndex90",
      ref,
      style: __console_patch ? { zIndex: 1999 } : void 0
    }),
    children
  )));
});
ToastPortal.displayName = "ToastPortal";
if (false) {
  ToastPortal.propTypes = ToastPortalPropTypes;
}

// src/ToastContainer.tsx
var ToastContainer = React8.forwardRef((_a, ref) => {
  var _b = _a, { children } = _b, props = __objRest(_b, ["children"]);
  return /* @__PURE__ */ React8.createElement(ToastPortal, __spreadValues({}, props), /* @__PURE__ */ React8.createElement(import_stack.Stack, { orientation: "vertical", spacing: "space40", ref }, children));
});
ToastContainer.displayName = "ToastContainer";
ToastContainer.propTypes = ToastContainerPropTypes;

// src/Toaster.tsx
var React9 = __toESM(require("react"));
var import_animation_library = require("@twilio-paste/animation-library");
var import_theme2 = require("@twilio-paste/theme");
var import_box7 = require("@twilio-paste/box");
var AnimatedToast = (0, import_animation_library.animated)(import_box7.Box);
var Toaster = (_a) => {
  var _b = _a, { toasts, pop } = _b, props = __objRest(_b, ["toasts", "pop"]);
  const prefersReducedMotion = (0, import_animation_library.useReducedMotion)();
  const [refMap] = React9.useState(() => /* @__PURE__ */ new WeakMap());
  const [returnTarget, setReturnTarget] = React9.useState({ trigger: null });
  const theme = (0, import_theme2.useTheme)();
  const transitions = (0, import_animation_library.useTransition)(toasts, {
    from: {
      marginBottom: "0rem",
      opacity: 0,
      transform: "translateX(100%) scale(1)",
      height: prefersReducedMotion ? "auto" : 0
    },
    enter: prefersReducedMotion ? { marginBottom: theme.space.space40, opacity: 1, transform: "translateX(0px) scale(1)" } : (item) => async (next) => {
      await next({
        marginBottom: theme.space.space40,
        opacity: 1,
        transform: "translateX(0px) scale(1)",
        height: refMap.get(item).offsetHeight
      });
    },
    leave: {
      marginBottom: "0rem",
      opacity: 0,
      transform: "translateX(0px) scale(0.8)",
      height: prefersReducedMotion ? "auto" : 0
    },
    config: {
      mass: 1,
      tension: 150,
      friction: 20
    }
  });
  const clearReturnTarget = React9.useCallback(() => {
    setReturnTarget({ trigger: null });
  }, [returnTarget]);
  React9.useEffect(() => {
    if (returnTarget.trigger != null && toasts.length === 0) {
      returnTarget.trigger.focus();
      clearReturnTarget();
    }
  }, [toasts]);
  return /* @__PURE__ */ React9.createElement(ToastPortal, __spreadValues({}, props), transitions((styles, item, state, index) => {
    return /* @__PURE__ */ React9.createElement(AnimatedToast, { as: "div", key: state.key, style: styles }, /* @__PURE__ */ React9.createElement(
      Toast,
      __spreadProps(__spreadValues({
        key: item.id
      }, item), {
        onDismiss: () => {
          if (item.onDismiss != null) {
            item.onDismiss();
          }
          pop(item.id);
        },
        ref: (ref) => {
          if (ref != null) {
            refMap.set(item, ref);
          }
        },
        setFocus: index === 0,
        onFocus: (e) => {
          if (e.relatedTarget && !returnTarget.trigger) {
            setReturnTarget({ trigger: e.relatedTarget });
          }
        }
      }),
      item.message
    ));
  }));
};
Toaster.displayName = "Toaster";
if (false) {
  Toaster.propTypes = ToasterPropTypes;
}

// src/useToaster.ts
var React10 = __toESM(require("react"));
var import_uid_library = require("@twilio-paste/uid-library");
var useToaster = () => {
  const isMounted = React10.useRef(null);
  const [toasts, setToasts] = React10.useState([]);
  React10.useEffect(() => {
    isMounted.current = true;
    return () => {
      isMounted.current = false;
      toasts.forEach((toast) => {
        if (toast.timeOutId) {
          window.clearTimeout(toast.timeOutId);
        }
      });
    };
  }, []);
  const pop = (id) => {
    if (!isMounted.current) {
      return;
    }
    setToasts(
      (currentToasts) => currentToasts.filter((toast) => {
        if (toast.id === id) {
          if (toast.timeOutId) {
            window.clearTimeout(toast.timeOutId);
          }
          if (toast.onDismiss) {
            toast.onDismiss();
          }
        }
        return toast.id !== id;
      })
    );
  };
  const push = (newToast) => {
    if (!isMounted.current) {
      return;
    }
    const generatedID = (0, import_uid_library.uid)(newToast);
    let timeOutId;
    if (newToast.dismissAfter != null && Number.isInteger(newToast.dismissAfter)) {
      timeOutId = window.setTimeout(pop, newToast.dismissAfter, newToast.id || generatedID);
    }
    const existingToasts = toasts.map((toast) => {
      const tmpToast = toast;
      tmpToast.setFocus = false;
      return tmpToast;
    });
    setToasts([__spreadValues({ id: generatedID, timeOutId, setFocus: true }, newToast), ...existingToasts]);
  };
  return { toasts, push, pop };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AnimatedToast,
  Toast,
  ToastContainer,
  Toaster,
  useToaster
});
