// @twilio-paste/code-block/dist/index.debug.es.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};

// src/CodeBlock.tsx
import * as React3 from "react";
import PropTypes from "prop-types";
import { SyntaxHighlighter } from "@twilio-paste/syntax-highlighter-library";
import { Box as Box2, safelySpreadBoxProps } from "@twilio-paste/box";

// src/CodeBlockTheme.ts
import { css } from "@twilio-paste/styling-library";
import { useTheme } from "@twilio-paste/theme";
var getPasteSyntaxTheme = (maxLines) => {
  const theme = useTheme();
  return css({
    'code[class*="language-"]': {
      color: "#d6deeb",
      fontFamily: "fontFamilyCode",
      textAlign: "left",
      whiteSpace: "pre",
      wordSpacing: "normal",
      wordBreak: "normal",
      wordWrap: "normal",
      lineHeight: "1.5",
      fontSize: "1em",
      MozTabSize: "4",
      OTabSize: "4",
      tabSize: "4",
      WebkitHyphens: "none",
      MozHyphens: "none",
      msHyphens: "none",
      hyphens: "none"
    },
    'pre[class*="language-"]': {
      color: "white",
      fontFamily: "fontFamilyCode",
      textAlign: "left",
      whiteSpace: "pre",
      wordSpacing: "normal",
      wordBreak: "normal",
      wordWrap: "normal",
      lineHeight: "1.5",
      fontSize: "1em",
      MozTabSize: "4",
      OTabSize: "4",
      tabSize: "4",
      WebkitHyphens: "none",
      MozHyphens: "none",
      msHyphens: "none",
      hyphens: "none",
      margin: "0",
      overflow: "auto",
      background: "#121C2D",
      display: "-webkit-box",
      WebkitLineClamp: `${maxLines}`,
      WebkitBoxOrient: "vertical",
      width: "100%"
    },
    'pre[class*="language-"]::-moz-selection': {
      textShadow: "none",
      background: "rgba(29, 59, 83, 0.99)"
    },
    'pre[class*="language-"] ::-moz-selection': {
      textShadow: "none",
      background: "rgba(29, 59, 83, 0.99)"
    },
    'code[class*="language-"]::-moz-selection': {
      textShadow: "none",
      background: "rgba(29, 59, 83, 0.99)"
    },
    'code[class*="language-"] ::-moz-selection': {
      textShadow: "none",
      background: "rgba(29, 59, 83, 0.99)"
    },
    'pre[class*="language-"]::selection': {
      textShadow: "none",
      background: "rgba(29, 59, 83, 0.99)"
    },
    'pre[class*="language-"] ::selection': {
      textShadow: "none",
      background: "rgba(29, 59, 83, 0.99)"
    },
    'code[class*="language-"]::selection': {
      textShadow: "none",
      background: "rgba(29, 59, 83, 0.99)"
    },
    'code[class*="language-"] ::selection': {
      textShadow: "none",
      background: "rgba(29, 59, 83, 0.99)"
    },
    ':not(pre) > code[class*="language-"]': {
      color: "white",
      background: "#011627",
      padding: "0.1em",
      borderRadius: "0.3em",
      whiteSpace: "normal"
    },
    comment: {
      color: "rgb(136, 145, 170)",
      fontStyle: "italic"
    },
    prolog: {
      color: "rgb(99, 119, 119)",
      fontStyle: "italic"
    },
    cdata: {
      color: "rgb(99, 119, 119)",
      fontStyle: "italic"
    },
    punctuation: {
      color: "rgb(199, 146, 234)"
    },
    ".namespace": {
      color: "rgb(178, 204, 214)"
    },
    deleted: {
      color: "rgba(239, 83, 80, 0.56)",
      fontStyle: "italic"
    },
    symbol: {
      color: "rgb(128, 203, 196)"
    },
    property: {
      color: "rgb(128, 203, 196)"
    },
    tag: {
      color: "rgb(127, 219, 202)"
    },
    operator: {
      color: "rgb(127, 219, 202)"
    },
    keyword: {
      color: "rgb(127, 219, 202)"
    },
    boolean: {
      color: "rgb(255, 88, 116)"
    },
    number: {
      color: "rgb(247, 140, 108)"
    },
    constant: {
      color: "rgb(130, 170, 255)"
    },
    function: {
      color: "rgb(130, 170, 255)"
    },
    builtin: {
      color: "rgb(130, 170, 255)"
    },
    char: {
      color: "rgb(130, 170, 255)"
    },
    selector: {
      color: "rgb(199, 146, 234)",
      fontStyle: "italic"
    },
    doctype: {
      color: "rgb(199, 146, 234)",
      fontStyle: "italic"
    },
    "attr-name": {
      color: "rgb(173, 219, 103)",
      fontStyle: "italic"
    },
    inserted: {
      color: "rgb(173, 219, 103)",
      fontStyle: "italic"
    },
    string: {
      color: "rgb(173, 219, 103)"
    },
    url: {
      color: "rgb(173, 219, 103)"
    },
    entity: {
      color: "rgb(173, 219, 103)"
    },
    ".language-css .token.string": {
      color: "rgb(173, 219, 103)"
    },
    ".style .token.string": {
      color: "rgb(173, 219, 103)"
    },
    "class-name": {
      color: "rgb(255, 203, 139)"
    },
    atrule: {
      color: "rgb(255, 203, 139)"
    },
    "attr-value": {
      color: "rgb(255, 203, 139)"
    },
    regex: {
      color: "rgb(214, 222, 235)"
    },
    important: {
      color: "rgb(214, 222, 235)",
      fontWeight: "bold"
    },
    variable: {
      color: "rgb(214, 222, 235)"
    },
    bold: {
      fontWeight: "bold"
    },
    italic: {
      fontStyle: "italic"
    }
  })(theme);
};

// src/CopyButton.tsx
import * as React from "react";
import { Button } from "@twilio-paste/button";
import { Box } from "@twilio-paste/box";
import { useClipboard } from "@twilio-paste/clipboard-copy-library";
import { ScreenReaderOnly } from "@twilio-paste/screen-reader-only";
import { CopyIcon } from "@twilio-paste/icons/esm/CopyIcon";
import { Tooltip, useTooltipState } from "@twilio-paste/tooltip";
var getCopyButtonText = (labelBefore, labelAfter) => {
  return (copied) => {
    return copied ? labelAfter : labelBefore;
  };
};
var CopyButton = ({
  text,
  i18nCopyLabelBefore = "Copy code block",
  i18nCopyLabelAfter = "Copied!",
  element = "COPY_BUTTON",
  copyTextFormatter
}) => {
  const tooltipState = useTooltipState();
  const [tooltipText, setTooltipText] = React.useState(i18nCopyLabelBefore);
  const isFirstRender = React.useRef(true);
  const getText = React.useMemo(() => {
    return getCopyButtonText(i18nCopyLabelBefore, i18nCopyLabelAfter);
  }, [i18nCopyLabelBefore, i18nCopyLabelAfter]);
  const clipboard = useClipboard({ copiedTimeout: 1500 });
  const handleCopy = React.useCallback(() => {
    const formattedText = copyTextFormatter ? copyTextFormatter(text) : text;
    clipboard.copy(formattedText);
  }, [text]);
  React.useEffect(() => {
    setTooltipText(getText(clipboard.copied));
  }, [clipboard.copied]);
  React.useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
      return;
    }
    tooltipState.hide();
    setTimeout(() => tooltipState.show(), 0);
  }, [tooltipText]);
  return /* @__PURE__ */ React.createElement(Tooltip, { text: tooltipText, state: tooltipState, element: `${element}_TOOLTIP` }, /* @__PURE__ */ React.createElement(Button, { variant: "inverse", size: "icon_small", onClick: handleCopy, element }, /* @__PURE__ */ React.createElement(Box, { as: "span" }, /* @__PURE__ */ React.createElement(CopyIcon, { decorative: true }), /* @__PURE__ */ React.createElement(Box, { as: "span", "aria-live": "polite" }, /* @__PURE__ */ React.createElement(ScreenReaderOnly, null, tooltipText)))));
};
CopyButton.displayName = "CopyButton";

// src/ExternalLinkButton.tsx
import * as React2 from "react";
import { Button as Button2 } from "@twilio-paste/button";
import { ScreenReaderOnly as ScreenReaderOnly2 } from "@twilio-paste/screen-reader-only";
import { LinkExternalIcon } from "@twilio-paste/icons/esm/LinkExternalIcon";
import { Tooltip as Tooltip2, useTooltipState as useTooltipState2 } from "@twilio-paste/tooltip";
var ExternalLinkButton = ({
  i18nLinkLabel = "Open code block in new page",
  href,
  element = "EXTERNAL_LINK"
}) => {
  const tooltipState = useTooltipState2();
  const tooltipText = i18nLinkLabel;
  const isFirstRender = React2.useRef(true);
  React2.useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
      return;
    }
    tooltipState.hide();
    setTimeout(() => tooltipState.show(), 0);
  }, [tooltipText]);
  return /* @__PURE__ */ React2.createElement(Tooltip2, { text: tooltipText, element: `${element}_TOOLTIP` }, /* @__PURE__ */ React2.createElement(Button2, { as: "a", variant: "inverse", size: "icon_small", href, element, target: "_blank" }, /* @__PURE__ */ React2.createElement(LinkExternalIcon, { decorative: true }), /* @__PURE__ */ React2.createElement(ScreenReaderOnly2, null, i18nLinkLabel)));
};
ExternalLinkButton.displayName = "ExternalLinkButton";

// src/CodeBlock.tsx
var CodeBlockVariantStyles = {
  "single-line": {
    gridTemplateColumns: "1fr auto",
    gridTemplateAreas: '"code-block button-group"',
    columnGap: "space40"
  },
  "multi-line": {
    gridTemplateRows: "auto 1fr",
    gridTemplateAreas: '"button-group" "code-block"',
    rowGap: "space20"
  }
};
var CodeBlock = React3.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      element = "CODE_BLOCK",
      language,
      code,
      variant = "multi-line",
      showLineNumbers,
      wrapLines,
      maxLines,
      externalLink,
      i18nCopyLabelAfter,
      i18nCopyLabelBefore,
      i18nLinkLabel,
      copyTextFormatter
    } = _b, props = __objRest(_b, [
      "element",
      "language",
      "code",
      "variant",
      "showLineNumbers",
      "wrapLines",
      "maxLines",
      "externalLink",
      "i18nCopyLabelAfter",
      "i18nCopyLabelBefore",
      "i18nLinkLabel",
      "copyTextFormatter"
    ]);
    return /* @__PURE__ */ React3.createElement(
      Box2,
      __spreadValues(__spreadProps(__spreadValues({
        ref,
        element,
        variant
      }, safelySpreadBoxProps(props)), {
        backgroundColor: "colorBackgroundBodyInverse",
        padding: "space50",
        display: "grid"
      }), CodeBlockVariantStyles[variant]),
      /* @__PURE__ */ React3.createElement(
        Box2,
        {
          display: "flex",
          flexDirection: "row",
          alignItems: "center",
          columnGap: "space30",
          justifyContent: variant === "multi-line" ? "flex-end" : void 0,
          gridArea: "button-group"
        },
        /* @__PURE__ */ React3.createElement(Box2, null, /* @__PURE__ */ React3.createElement(
          CopyButton,
          {
            element: `${element}_COPY_BUTTON`,
            text: code,
            i18nCopyLabelAfter,
            i18nCopyLabelBefore,
            copyTextFormatter
          }
        )),
        externalLink && /* @__PURE__ */ React3.createElement(Box2, null, /* @__PURE__ */ React3.createElement(
          ExternalLinkButton,
          {
            href: externalLink,
            element: `${element}_EXTERNAL_LINK`,
            i18nLinkLabel
          }
        ))
      ),
      /* @__PURE__ */ React3.createElement(Box2, { element: `${element}_CONTENT`, display: "flex", alignItems: "center", overflowX: "auto", gridArea: "code-block" }, /* @__PURE__ */ React3.createElement(
        SyntaxHighlighter,
        {
          language,
          style: getPasteSyntaxTheme(maxLines),
          showLineNumbers,
          wrapLongLines: wrapLines,
          wrapLines: true
        },
        code
      ))
    );
  }
);
CodeBlock.displayName = "CodeBlock";
CodeBlock.propTypes = {
  code: PropTypes.string.isRequired,
  element: PropTypes.string,
  language: PropTypes.oneOf([
    "javascript",
    "jsx",
    "csharp",
    "php",
    "ruby",
    "python",
    "java",
    "json",
    "c",
    "bash",
    "shell-session",
    "go",
    "groovy"
  ]).isRequired,
  showLineNumbers: PropTypes.bool,
  wrapLines: PropTypes.bool,
  maxLines: PropTypes.number,
  externalLink: PropTypes.string,
  i18nCopyLabelBefore: PropTypes.string,
  i18nCopyLabelAfter: PropTypes.string,
  i18nLinkLabel: PropTypes.string,
  variant: PropTypes.oneOf(["multi-line", "single-line"]),
  copyTextFormatter: PropTypes.func
};

// src/CodeBlockHeader.tsx
import * as React4 from "react";
import { Box as Box3, safelySpreadBoxProps as safelySpreadBoxProps2 } from "@twilio-paste/box";
var CodeBlockHeader = React4.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "CODE_BLOCK_HEADER", as = "h3" } = _b, props = __objRest(_b, ["children", "element", "as"]);
    return /* @__PURE__ */ React4.createElement(
      Box3,
      __spreadProps(__spreadValues({}, safelySpreadBoxProps2(props)), {
        ref,
        element,
        as,
        backgroundColor: "colorBackgroundInverse",
        color: "colorTextInverse",
        paddingY: "space40",
        paddingX: "space60",
        margin: "space0",
        borderTopLeftRadius: "borderRadius20",
        borderTopRightRadius: "borderRadius20",
        fontWeight: "fontWeightSemibold",
        fontSize: "fontSize30"
      }),
      children
    );
  }
);
CodeBlockHeader.displayName = "CodeBlockHeader";

// src/CodeBlockTab.tsx
import * as React6 from "react";
import PropTypes2 from "prop-types";
import { Box as Box4, safelySpreadBoxProps as safelySpreadBoxProps3 } from "@twilio-paste/box";
import { TabPrimitive } from "@twilio-paste/tabs-primitive";

// src/CodeBlockTabsContext.tsx
import * as React5 from "react";
var CodeBlockTabsContext = React5.createContext({});

// src/CodeBlockTab.tsx
var StyledCodeBlockTab = React6.forwardRef((_a, ref) => {
  var _b = _a, { children } = _b, props = __objRest(_b, ["children"]);
  return /* @__PURE__ */ React6.createElement(
    Box4,
    __spreadProps(__spreadValues({
      ref,
      as: "button"
    }, safelySpreadBoxProps3(props)), {
      transition: "background-color 150ms ease-in, color 150ms ease-in, border-color 150ms ease-in, box-shadow 150ms ease-in",
      backgroundColor: "colorBackgroundInverse",
      color: "colorTextInverseWeak",
      fontSize: "fontSize30",
      lineHeight: "lineHeight30",
      paddingY: "space30",
      paddingX: "space40",
      margin: "space0",
      borderStyle: "solid",
      borderWidth: "borderWidth10",
      borderColor: "transparent",
      borderRadius: "borderRadius20",
      _focus: { outline: "none", boxShadow: "shadowFocusInverse" },
      _hover: { backgroundColor: "colorBackgroundPrimary", color: "colorTextInverse" },
      _selected: {
        backgroundColor: "colorBackgroundBodyInverse",
        color: "colorTextBrandInverse",
        borderColor: "colorBorderInverseWeaker"
      },
      _selected_focus: { borderColor: "transparent" },
      _selected_hover: {
        backgroundColor: "colorBackgroundPrimaryStronger",
        color: "colorTextWeakest",
        borderColor: "transparent"
      }
    }),
    children
  );
});
StyledCodeBlockTab.displayName = "StyledCodeBlockTab";
var CodeBlockTab = React6.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "CODE_BLOCK_TAB" } = _b, props = __objRest(_b, ["children", "element"]);
    const tabState = React6.useContext(CodeBlockTabsContext);
    return /* @__PURE__ */ React6.createElement(TabPrimitive, __spreadValues(__spreadProps(__spreadValues({}, tabState), { as: StyledCodeBlockTab, element, ref }), props), children);
  }
);
CodeBlockTab.displayName = "CodeBlockTab";
CodeBlockTab.propTypes = {
  children: PropTypes2.node,
  element: PropTypes2.string
};

// src/CodeBlockTabList.tsx
import * as React7 from "react";
import PropTypes3 from "prop-types";
import { Box as Box5, safelySpreadBoxProps as safelySpreadBoxProps4 } from "@twilio-paste/box";
import { TabPrimitiveList } from "@twilio-paste/tabs-primitive";
var CodeBlockTabList = React7.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "CODE_BLOCK_TAB_LIST" } = _b, props = __objRest(_b, ["children", "element"]);
    const tabState = React7.useContext(CodeBlockTabsContext);
    return /* @__PURE__ */ React7.createElement(
      TabPrimitiveList,
      __spreadProps(__spreadValues(__spreadValues({}, safelySpreadBoxProps4(props)), tabState), {
        as: Box5,
        ref,
        element,
        display: "flex",
        columnGap: "space50",
        backgroundColor: "colorBackgroundInverse",
        paddingBottom: "space40",
        paddingX: "space60"
      }),
      children
    );
  }
);
CodeBlockTabList.displayName = "CodeBlockTabList";
CodeBlockTabList.propTypes = {
  children: PropTypes3.node,
  element: PropTypes3.string
};

// src/CodeBlockTabPanel.tsx
import * as React8 from "react";
import PropTypes4 from "prop-types";
import { Box as Box6, safelySpreadBoxProps as safelySpreadBoxProps5 } from "@twilio-paste/box";
import { TabPrimitivePanel } from "@twilio-paste/tabs-primitive";
var CodeBlockTabPanel = React8.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, element = "CODE_BLOCK_TAB_PANEL" } = _b, props = __objRest(_b, ["children", "element"]);
    const tabState = React8.useContext(CodeBlockTabsContext);
    return /* @__PURE__ */ React8.createElement(TabPrimitivePanel, __spreadProps(__spreadValues(__spreadValues({}, safelySpreadBoxProps5(props)), tabState), { as: Box6, element, ref }), children);
  }
);
CodeBlockTabPanel.displayName = "CodeBlockTabPanel";
CodeBlockTabPanel.propTypes = {
  children: PropTypes4.node,
  element: PropTypes4.string
};

// src/CodeBlockWrapper.tsx
import * as React9 from "react";
import PropTypes5 from "prop-types";
import { Box as Box7 } from "@twilio-paste/box";
import { useTabPrimitiveState } from "@twilio-paste/tabs-primitive";
var CodeBlockWrapper = React9.forwardRef(
  (_a, ref) => {
    var _b = _a, { children, state, element = "CODE_BLOCK_WRAPPER" } = _b, initialState = __objRest(_b, ["children", "state", "element"]);
    const tabState = state || useTabPrimitiveState(__spreadValues({ orientation: "horizontal" }, initialState));
    return /* @__PURE__ */ React9.createElement(CodeBlockTabsContext.Provider, { value: tabState }, /* @__PURE__ */ React9.createElement(Box7, { element, ref }, children));
  }
);
CodeBlockWrapper.displayName = "CodeBlockWrapper";
CodeBlockWrapper.propTypes = {
  children: PropTypes5.node,
  element: PropTypes5.string
};
export {
  CodeBlock,
  CodeBlockHeader,
  CodeBlockTab,
  CodeBlockTabList,
  CodeBlockTabPanel,
  CodeBlockWrapper
};
