// @twilio-paste/textarea/dist/index.debug.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  TextArea: () => TextArea
});
module.exports = __toCommonJS(src_exports);

// src/TextArea.tsx
var React = __toESM(require("react"));
var import_prop_types = __toESM(require("prop-types"));
var import_react_textarea_autosize_library = __toESM(require("@twilio-paste/react-textarea-autosize-library"));
var import_styling_library = require("@twilio-paste/styling-library");
var import_box = require("@twilio-paste/box");
var import_input_box = require("@twilio-paste/input-box");
var TextAreaElement = (0, import_styling_library.styled)(import_react_textarea_autosize_library.default)(
  (props) => (0, import_styling_library.css)({
    appearance: "none",
    background: "transparent",
    border: "none",
    borderRadius: "borderRadius20",
    boxShadow: "none",
    color: "inherit",
    display: "block",
    fontFamily: "inherit",
    fontSize: "fontSize30",
    fontWeight: "fontWeightMedium",
    lineHeight: "lineHeight20",
    maxHeight: props.resize === "vertical" ? "none" : "size30",
    outline: "none",
    paddingBottom: "space30",
    paddingLeft: "space40",
    paddingRight: "space40",
    paddingTop: "space30",
    resize: props.resize,
    width: "100%",
    "&::placeholder": {
      color: props.variant === "inverse" ? "colorTextInverseWeak" : "colorTextWeak",
      fontStyle: "italic"
    },
    "&:focus::placeholder": {
      color: props.variant === "inverse" ? "colorTextInverseWeak" : "colorTextWeak"
    },
    "&:disabled": {
      color: props.variant === "inverse" ? "colorTextInverseWeaker" : "colorTextWeaker",
      cursor: "not-allowed",
      "-webkit-text-fill-color": props.variant === "inverse" ? "colorTextInverseWeaker" : "colorTextWeaker",
      "-webkit-opacity": "1"
    }
  }),
  import_box.getCustomElementStyles
);
var TextArea = React.forwardRef(
  (_a, ref) => {
    var _b = _a, {
      children,
      disabled,
      element = "TEXTAREA",
      hasError,
      insertBefore,
      insertAfter,
      readOnly,
      variant,
      resize = "none",
      size,
      height,
      width
    } = _b, props = __objRest(_b, [
      "children",
      "disabled",
      "element",
      "hasError",
      "insertBefore",
      "insertAfter",
      "readOnly",
      "variant",
      "resize",
      "size",
      "height",
      "width"
    ]);
    return /* @__PURE__ */ React.createElement(
      import_input_box.InputBox,
      {
        disabled,
        element,
        hasError,
        insertAfter,
        insertBefore,
        readOnly,
        variant
      },
      /* @__PURE__ */ React.createElement(
        TextAreaElement,
        __spreadProps(__spreadValues({}, (0, import_box.safelySpreadBoxProps)(props)), {
          "aria-invalid": hasError,
          "aria-readonly": readOnly,
          disabled,
          "data-paste-element": `${element}_ELEMENT`,
          readOnly,
          ref,
          rows: 3,
          minRows: 3,
          spellCheck: true,
          resize,
          variant
        }),
        children
      )
    );
  }
);
TextArea.displayName = "TextArea";
TextArea.propTypes = {
  disabled: import_prop_types.default.bool,
  element: import_prop_types.default.string,
  hasError: import_prop_types.default.bool,
  id: import_prop_types.default.string,
  name: import_prop_types.default.string,
  onBlur: import_prop_types.default.func,
  onChange: import_prop_types.default.func,
  onFocus: import_prop_types.default.func,
  placeholder: import_prop_types.default.string,
  readOnly: import_prop_types.default.bool,
  required: import_prop_types.default.bool
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TextArea
});
